// @flowglad/shared/dist/esm/utils.mjs
import "./chunk-2MM5EJJX.mjs";
const IS_DEV = process.env.NODE_ENV === "development";
const getBaseURL = () => {
  if (IS_DEV && process.env.FLOWGLAD_API_URL_OVERRIDE) {
    return process.env.FLOWGLAD_API_URL_OVERRIDE;
  }
  return "https://app.flowglad.com";
};
const constructCheckFeatureAccess = (subscriptions) => {
  const checkFeatureAccess = (featureSlug, refinementParams) => {
    var _a;
    const subscription = (refinementParams == null ? void 0 : refinementParams.subscriptionId) ? subscriptions.find(
      (s) => s.id === refinementParams.subscriptionId
    ) : subscriptions[0];
    if (!subscription) {
      return false;
    }
    const experimental = subscription.experimental;
    const featureItemsBySlug = (_a = experimental == null ? void 0 : experimental.featureItems.reduce(
      (acc, featureItem2) => {
        if (featureItem2.type === "toggle") {
          acc[featureItem2.slug] = featureItem2;
        }
        return acc;
      },
      {}
    )) != null ? _a : {};
    const featureItem = featureItemsBySlug[featureSlug];
    if (!featureItem) {
      return false;
    }
    return featureItem.type === "toggle";
  };
  return checkFeatureAccess;
};
const constructCheckUsageBalance = (subscriptions) => {
  const checkUsageBalance = (usageMeterSlug, refinementParams) => {
    var _a;
    const subscription = (refinementParams == null ? void 0 : refinementParams.subscriptionId) ? subscriptions.find(
      (s) => s.id === refinementParams.subscriptionId
    ) : subscriptions[0];
    if (!subscription) {
      return null;
    }
    const experimental = subscription.experimental;
    const usageMeterBalancesBySlug = (_a = experimental == null ? void 0 : experimental.usageMeterBalances.reduce(
      (acc, usageMeterBalance2) => {
        acc[usageMeterBalance2.slug] = usageMeterBalance2;
        return acc;
      },
      {}
    )) != null ? _a : {};
    const usageMeterBalance = usageMeterBalancesBySlug[usageMeterSlug];
    if (!usageMeterBalance) {
      return null;
    }
    return usageMeterBalance;
  };
  return checkUsageBalance;
};
const constructGetProduct = (pricingModel) => {
  const productsBySlug = new Map(
    pricingModel.products.map((product) => [product.slug, product])
  );
  const getProduct = (productSlug) => {
    var _a;
    return (_a = productsBySlug.get(productSlug)) != null ? _a : null;
  };
  return getProduct;
};
const constructGetPrice = (pricingModel) => {
  const allPrices = pricingModel.products.flatMap(
    (product) => product.prices.map((price) => [price.slug, price])
  );
  const pricesBySlug = new Map(allPrices);
  const getPrice = (priceSlug) => {
    var _a;
    return (_a = pricesBySlug.get(priceSlug)) != null ? _a : null;
  };
  return getPrice;
};
const constructHasPurchased = (pricingModel, purchases) => {
  const productsBySlug = new Map(
    pricingModel.products.map((product) => [product.slug, product])
  );
  const purchasedPriceIds = new Set(
    (purchases != null ? purchases : []).map((purchase) => purchase.priceId)
  );
  const hasPurchased = (productSlug) => {
    const product = productsBySlug.get(productSlug);
    if (!product) {
      return false;
    }
    return product.prices.some(
      (price) => purchasedPriceIds.has(price.id)
    );
  };
  return hasPurchased;
};
export {
  constructCheckFeatureAccess,
  constructCheckUsageBalance,
  constructGetPrice,
  constructGetProduct,
  constructHasPurchased,
  getBaseURL
};
