// @flowglad/server/dist/esm/better-auth/constants.mjs
import "../chunk-E3LOUS7X.mjs";
import {
  FlowgladActionKey
} from "@flowglad/shared";
const endpointKeyToActionKey = {
  getCustomerBilling: FlowgladActionKey.GetCustomerBilling,
  findOrCreateCustomer: FlowgladActionKey.FindOrCreateCustomer,
  createCheckoutSession: FlowgladActionKey.CreateCheckoutSession,
  createAddPaymentMethodCheckoutSession: FlowgladActionKey.CreateAddPaymentMethodCheckoutSession,
  createActivateSubscriptionCheckoutSession: FlowgladActionKey.CreateActivateSubscriptionCheckoutSession,
  cancelSubscription: FlowgladActionKey.CancelSubscription,
  uncancelSubscription: FlowgladActionKey.UncancelSubscription,
  adjustSubscription: FlowgladActionKey.AdjustSubscription,
  createSubscription: FlowgladActionKey.CreateSubscription,
  updateCustomer: FlowgladActionKey.UpdateCustomer,
  createUsageEvent: FlowgladActionKey.CreateUsageEvent,
  getResources: FlowgladActionKey.GetResourceUsages,
  getResourceUsage: FlowgladActionKey.GetResourceUsage,
  claimResource: FlowgladActionKey.ClaimResource,
  releaseResource: FlowgladActionKey.ReleaseResource,
  listResourceClaims: FlowgladActionKey.ListResourceClaims,
  getUsageMeterBalances: FlowgladActionKey.GetUsageMeterBalances
};
const _actionKeyToEndpointKey = {
  [FlowgladActionKey.GetCustomerBilling]: "getCustomerBilling",
  [FlowgladActionKey.FindOrCreateCustomer]: "findOrCreateCustomer",
  [FlowgladActionKey.CreateCheckoutSession]: "createCheckoutSession",
  [FlowgladActionKey.CreateAddPaymentMethodCheckoutSession]: "createAddPaymentMethodCheckoutSession",
  [FlowgladActionKey.CreateActivateSubscriptionCheckoutSession]: "createActivateSubscriptionCheckoutSession",
  [FlowgladActionKey.CancelSubscription]: "cancelSubscription",
  [FlowgladActionKey.UncancelSubscription]: "uncancelSubscription",
  [FlowgladActionKey.AdjustSubscription]: "adjustSubscription",
  [FlowgladActionKey.CreateSubscription]: "createSubscription",
  [FlowgladActionKey.UpdateCustomer]: "updateCustomer",
  [FlowgladActionKey.CreateUsageEvent]: "createUsageEvent",
  [FlowgladActionKey.GetResourceUsages]: "getResources",
  [FlowgladActionKey.GetResourceUsage]: "getResourceUsage",
  [FlowgladActionKey.ClaimResource]: "claimResource",
  [FlowgladActionKey.ReleaseResource]: "releaseResource",
  [FlowgladActionKey.ListResourceClaims]: "listResourceClaims",
  [FlowgladActionKey.GetUsageMeterBalances]: "getUsageMeterBalances"
};
const _hybridActionKeyToEndpointKey = {
  [FlowgladActionKey.GetPricingModel]: "getPricingModel"
};
export {
  endpointKeyToActionKey
};
