// @flowglad/server/dist/esm/better-auth/plugin.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "../chunk-E3LOUS7X.mjs";
import { createAuthMiddleware } from "better-auth/plugins";
import { FlowgladServer } from "../FlowgladServer.mjs";
import {
  createBillingEndpoints,
  createGetExternalIdEndpoint,
  createGetPricingModelEndpoint,
  getCreatorRoleFromOrgOptions,
  getOrganizationDetails,
  getOrgOptionsFromCtxContext,
  isAdapterLike
} from "./endpoints.mjs";
const getDefaultCustomer = (session, customerType) => __async(null, null, function* () {
  if (customerType === "organization") {
    const orgId = session.user.organizationId;
    if (!orgId) {
      throw new Error("Organization ID not found in session");
    }
    return {
      externalId: orgId,
      name: session.user.name || "Organization",
      email: session.user.email || ""
    };
  }
  if (!session.user.email) {
    throw new Error(
      "User email is required to create Flowglad customer"
    );
  }
  return {
    externalId: session.user.id,
    name: session.user.name || "",
    email: session.user.email
  };
});
const createFlowgladCustomer = (options, session) => __async(null, null, function* () {
  const customerType = options.customerType || "user";
  const customerInfo = options.getCustomer ? yield options.getCustomer(session) : yield getDefaultCustomer(session, customerType);
  const apiKey = options.apiKey || process.env.FLOWGLAD_SECRET_KEY;
  const flowgladServerConfig = {
    customerExternalId: customerInfo.externalId,
    getCustomerDetails: () => __async(null, null, function* () {
      return {
        name: customerInfo.name,
        email: customerInfo.email
      };
    })
  };
  if (apiKey) {
    flowgladServerConfig.apiKey = apiKey;
  }
  if (options.baseURL) {
    flowgladServerConfig.baseURL = options.baseURL;
  }
  const flowgladServer = new FlowgladServer(flowgladServerConfig);
  yield flowgladServer.findOrCreateCustomer();
});
const reportCustomerCreateError = (options, params) => __async(null, null, function* () {
  if (options.onCustomerCreateError) {
    yield options.onCustomerCreateError(params);
    return;
  }
  if (params.hook === "afterSignUp") {
    console.error(
      "Failed to create Flowglad customer after sign-up:",
      params.error
    );
    return;
  }
  console.error(
    "Failed to create Flowglad customer after organization creation:",
    params.error
  );
});
const createFlowgladCustomerForOrganization = (options, params) => __async(null, null, function* () {
  var _a;
  const {
    organizationId,
    userId,
    userEmail,
    userName,
    organizationName,
    organizationEmail
  } = params;
  const session = {
    user: {
      id: userId,
      email: organizationEmail != null ? organizationEmail : userEmail,
      name: (_a = organizationName != null ? organizationName : userName) != null ? _a : null,
      organizationId
    }
  };
  yield createFlowgladCustomer(options, session);
});
const flowgladPlugin = (options) => {
  const plugin = {
    id: "flowglad",
    endpoints: __spreadProps(__spreadValues({
      getExternalId: createGetExternalIdEndpoint(options)
    }, createBillingEndpoints(options)), {
      getPricingModel: createGetPricingModelEndpoint(options)
    }),
    hooks: {
      after: [
        {
          matcher: (context) => {
            return context.path === "/sign-up" || context.path === "/sign-up/email";
          },
          handler: createAuthMiddleware((ctx) => __async(null, null, function* () {
            var _a;
            const returned = ctx.context.returned;
            if (!((_a = returned == null ? void 0 : returned.user) == null ? void 0 : _a.id)) {
              return;
            }
            const session = {
              user: {
                id: returned.user.id,
                name: returned.user.name || null,
                email: returned.user.email || null,
                organizationId: null
              }
            };
            const customerType = options.customerType || "user";
            if (customerType !== "user") {
              return;
            }
            try {
              yield createFlowgladCustomer(options, session);
            } catch (error) {
              yield reportCustomerCreateError(options, {
                hook: "afterSignUp",
                customerType,
                session,
                error
              });
            }
          }))
        },
        {
          matcher: (context) => {
            return context.path === "/organization/create";
          },
          handler: createAuthMiddleware((ctx) => __async(null, null, function* () {
            var _a, _b, _c, _d, _e;
            const returned = ctx.context.returned;
            if (!((_a = returned == null ? void 0 : returned.organization) == null ? void 0 : _a.id)) {
              return;
            }
            const session = ctx.context.session;
            if (!(session == null ? void 0 : session.user)) {
              return;
            }
            const customerType = options.customerType || "user";
            if (customerType !== "organization") {
              return;
            }
            const organizationId = returned.organization.id;
            const organizationName = typeof returned.organization.name === "string" ? returned.organization.name : "Organization";
            const organizationEmail = isAdapterLike(ctx.context.adapter) && typeof session.user.id === "string" ? (_d = (_c = (_b = yield getOrganizationDetails({
              adapter: ctx.context.adapter,
              organizationId,
              userId: session.user.id,
              creatorRole: getCreatorRoleFromOrgOptions(
                getOrgOptionsFromCtxContext(ctx.context)
              )
            })) == null ? void 0 : _b.email) != null ? _c : session.user.email) != null ? _d : null : (_e = session.user.email) != null ? _e : null;
            const orgSession = {
              user: {
                id: session.user.id,
                // For org customers, we want Flowglad customer name/email to
                // default to org name + owner email (best-effort).
                name: organizationName,
                email: organizationEmail,
                organizationId
              }
            };
            try {
              yield createFlowgladCustomer(options, orgSession);
            } catch (error) {
              yield reportCustomerCreateError(options, {
                hook: "afterOrganizationCreate",
                customerType,
                session: orgSession,
                error
              });
            }
          }))
        }
      ]
    }
  };
  return plugin;
};
export {
  createFlowgladCustomerForOrganization,
  flowgladPlugin
};
