// @flowglad/react/dist/cjs/lib/utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  cn: () => cn,
  currencyCharacter: () => currencyCharacter,
  devError: () => devError,
  devWarn: () => devWarn,
  formatDate: () => formatDate,
  humanReadableCurrencyAmount: () => humanReadableCurrencyAmount,
  isCurrencyZeroDecimal: () => isCurrencyZeroDecimal,
  zeroDecimalCurrencies: () => zeroDecimalCurrencies
});
module.exports = __toCommonJS(utils_exports);
var import_clsx = require("clsx");
var import_date_fns = require("date-fns");
var import_tailwind_merge = require("tailwind-merge");
function cn(...inputs) {
  return (0, import_tailwind_merge.twMerge)((0, import_clsx.clsx)(inputs));
}
function currencyCharacter(currency, locale) {
  const defaultLocale = typeof navigator !== "undefined" ? navigator.language : "en-US";
  return 0 .toLocaleString(locale != null ? locale : defaultLocale, {
    style: "currency",
    currency,
    minimumFractionDigits: 0,
    maximumFractionDigits: 0
  }).replace(/\d/g, "").trim();
}
function humanReadableCurrencyAmount(currency, amount) {
  const formatter = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency
  });
  if (!isCurrencyZeroDecimal(currency)) {
    return formatter.format(Number((amount / 100).toFixed(2)));
  }
  return formatter.format(amount);
}
const zeroDecimalCurrencies = [
  "BIF",
  "CLP",
  "DJF",
  "GNF",
  "JPY",
  "KMF",
  "KRW",
  "MGA",
  "PYG",
  "RWF",
  "UGX",
  "VND",
  "VUV",
  "XAF",
  "XOF",
  "XPF"
];
function isCurrencyZeroDecimal(currency) {
  return zeroDecimalCurrencies.includes(currency);
}
function formatDate(date) {
  return (0, import_date_fns.format)(
    date instanceof Date ? date : new Date(date),
    "MMM d, yyyy"
  );
}
function devWarn(message) {
  if (process.env.NODE_ENV !== "production") {
    console.warn(`[flowglad]: ${message}`);
  }
}
function devError(message) {
  if (process.env.NODE_ENV !== "production") {
    console.error(`[flowglad]: ${message}`);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  cn,
  currencyCharacter,
  devError,
  devWarn,
  formatDate,
  humanReadableCurrencyAmount,
  isCurrencyZeroDecimal,
  zeroDecimalCurrencies
});
