// @flowglad/nextjs/dist/esm/nextRouteHandler.js
import {
  __async
} from "./chunk-E2434BE5.js";
import { requestHandler } from "@flowglad/server";
import { NextResponse } from "next/server";
const nextRouteHandler = (options) => {
  const {
    getCustomerExternalId,
    flowglad,
    onError,
    beforeRequest,
    afterRequest,
    baseURL
  } = options;
  const routeHandler = (_0, _1) => __async(null, [_0, _1], function* (req, {
    params
  }) {
    try {
      const handler = requestHandler({
        getCustomerExternalId,
        flowglad,
        onError,
        beforeRequest,
        afterRequest,
        baseURL
      });
      const resolvedParams = "then" in params ? yield params : params;
      const { path } = resolvedParams;
      const result = yield handler(
        {
          path,
          method: req.method,
          query: req.method === "GET" ? Object.fromEntries(req.nextUrl.searchParams) : void 0,
          body: req.method !== "GET" ? yield req.json().catch(() => ({})) : void 0
        },
        req
      );
      return NextResponse.json(
        {
          error: result.error,
          data: result.data
        },
        {
          status: result.status
        }
      );
    } catch (error) {
      if (onError) {
        onError(error);
      }
      return NextResponse.json(
        {
          error: {
            message: error instanceof Error ? error.message : "Internal server error"
          },
          data: null
        },
        { status: 500 }
      );
    }
  });
  return {
    GET: routeHandler,
    POST: routeHandler
  };
};
export {
  nextRouteHandler
};
