// @flowglad/nextjs/dist/esm/withMCPToolAuthorization.js
import {
  __async
} from "./chunk-E2434BE5.js";
import { createMcpHandler } from "@vercel/mcp-adapter";
const toolWithFeatureAccessCheck = (toolCallback, params) => {
  const wrappedCallback = (...args) => __async(null, null, function* () {
    const { featureSlug, flowgladServer, upgradePriceSlug } = params;
    const billing = yield flowgladServer.getBilling();
    const feature = billing.checkFeatureAccess(featureSlug);
    if (!feature) {
      const price = billing.getPrice(upgradePriceSlug);
      if (!price) {
        return {
          content: [
            {
              type: "text",
              text: "Upgrade price not found"
            }
          ]
        };
      }
      const checkoutSession = yield params.flowgladServer.createCheckoutSession({
        priceId: price.id,
        quantity: 1,
        successUrl: params.successUrl,
        cancelUrl: params.cancelUrl
      });
      return {
        content: [
          {
            type: "text",
            text: `Feature access denied. To access this feature, please upgrade at: ${checkoutSession.url}`
          }
        ]
      };
    }
    return yield toolCallback(...args);
  });
  return wrappedCallback;
};
const toolWithUsageBalanceCheck = (toolCallback, featureSlug) => {
  return (...args) => __async(null, null, function* () {
    return yield toolCallback(...args);
  });
};
const mcpHandlerWithFlowglad = (constructor, flowgladConstructor) => {
  const handler = (request) => __async(null, null, function* () {
    const flowglad = yield flowgladConstructor(request);
    const mcpHandler = createMcpHandler((server) => __async(null, null, function* () {
      yield constructor(server, flowglad);
      return server;
    }));
    return yield mcpHandler(request);
  });
  return handler;
};
export {
  mcpHandlerWithFlowglad,
  toolWithFeatureAccessCheck,
  toolWithUsageBalanceCheck
};
