// @vercel/mcp-adapter/dist/cli/index.js
#!/usr/bin/env node
'use strict';

var chunkUILOG6SR_js = require('../chunk-UILOG6SR.js');
var commander = require('commander');
var fs = require('fs/promises');
var path = require('path');
var chalk = require('chalk');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefault(fs);
var path__default = /*#__PURE__*/_interopDefault(path);
var chalk__default = /*#__PURE__*/_interopDefault(chalk);

var require_cli = chunkUILOG6SR_js.__commonJS({
  "src/cli/index.ts"() {
    var program = new commander.Command();
    var ROUTE_TEMPLATE = `import { createMcpHandler } from '@vercel/mcp-adapter';
import { z } from 'zod';

const handler = createMcpHandler(
  server => {
    server.tool(
      'roll_dice',
      'Rolls an N-sided die',
      { 
        sides: z.number().int().min(2)
      },
      async ({ sides }) => {
        const value = 1 + Math.floor(Math.random() * sides);
        return {
          content: [{ type: 'text', text: \`\u{1F3B2} You rolled a \${value}!\` }],
        };
      }
    );
  },
  {
    // Optional server options
  },
  {
    // Optional redis config
    redisUrl: process.env.REDIS_URL,
    // You need these endpoints
    basePath: '/api',
    maxDuration: 60,
    verboseLogs: true,
  }
);

export { handler as GET, handler as POST };
`;
    function detectPackageManager() {
      return chunkUILOG6SR_js.__async(this, null, function* () {
        const cwd = process.cwd();
        try {
          const files = yield fs__default.default.readdir(cwd);
          if (files.includes("bun.lockb"))
            return "bun";
          if (files.includes("pnpm-lock.yaml"))
            return "pnpm";
          if (files.includes("yarn.lock"))
            return "yarn";
          if (files.includes("package-lock.json"))
            return "npm";
          return "npm";
        } catch (error) {
          return "npm";
        }
      });
    }
    function installDependencies(packageManager) {
      return chunkUILOG6SR_js.__async(this, null, function* () {
        const execSync = (yield import('child_process')).execSync;
        const dependencies = ["@vercel/mcp-adapter", "zod"];
        const commands = {
          npm: `npm install ${dependencies.join(" ")}`,
          pnpm: `pnpm add ${dependencies.join(" ")}`,
          yarn: `yarn add ${dependencies.join(" ")}`,
          bun: `bun add ${dependencies.join(" ")}`
        };
        try {
          console.log(
            chalk__default.default.blue(`
Installing dependencies using ${packageManager}...`)
          );
          execSync(commands[packageManager], { stdio: "inherit" });
          console.log(chalk__default.default.green("\n\u2705 Dependencies installed successfully!"));
        } catch (error) {
          console.error(
            chalk__default.default.red(
              "\n\u274C Failed to install dependencies. You can install them manually:"
            )
          );
          console.log(chalk__default.default.yellow(`
${commands[packageManager]}`));
        }
      });
    }
    function init() {
      return chunkUILOG6SR_js.__async(this, null, function* () {
        var _a, _b;
        try {
          const packageJson = JSON.parse(
            yield fs__default.default.readFile(path__default.default.join(process.cwd(), "package.json"), "utf-8")
          );
          if (!((_a = packageJson.dependencies) == null ? void 0 : _a.next) && !((_b = packageJson.devDependencies) == null ? void 0 : _b.next)) {
            console.error(
              chalk__default.default.red("\u274C This command must be run in a Next.js project")
            );
            process.exit(1);
          }
          const routePath = path__default.default.join(process.cwd(), "app", "api", "[transport]");
          yield fs__default.default.mkdir(routePath, { recursive: true });
          const routeFilePath = path__default.default.join(routePath, "route.ts");
          yield fs__default.default.writeFile(routeFilePath, ROUTE_TEMPLATE);
          console.log(chalk__default.default.green("\u2705 Successfully created MCP route handler!"));
          const packageManager = yield detectPackageManager();
          yield installDependencies(packageManager);
        } catch (error) {
          console.error(chalk__default.default.red("Error creating MCP route handler:"), error);
          process.exit(1);
        }
      });
    }
    program.name("@vercel/mcp-adapter").description("Initialize MCP route handler in your Next.js project").action(init);
    program.parse();
  }
});
var index = require_cli();

module.exports = index;
