// @flowglad/server/dist/cjs/subrouteHandlers/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var subrouteHandlers_exports = {};
__export(subrouteHandlers_exports, {
  hybridRouteToHandlerMap: () => hybridRouteToHandlerMap,
  isHybridActionKey: () => isHybridActionKey,
  routeToHandlerMap: () => routeToHandlerMap
});
module.exports = __toCommonJS(subrouteHandlers_exports);
var import_shared = require("@flowglad/shared");
var import_checkoutSessionHandlers = require('./checkoutSessionHandlers.cjs');
var import_customerHandlers = require('./customerHandlers.cjs');
var import_pricingModelHandlers = require('./pricingModelHandlers.cjs');
var import_resourceHandlers = require('./resourceHandlers.cjs');
var import_subscriptionHandlers = require('./subscriptionHandlers.cjs');
var import_usageEventHandlers = require('./usageEventHandlers.cjs');
var import_usageMeterHandlers = require('./usageMeterHandlers.cjs');
const routeToHandlerMap = {
  [import_shared.FlowgladActionKey.GetCustomerBilling]: import_customerHandlers.getCustomerBilling,
  [import_shared.FlowgladActionKey.FindOrCreateCustomer]: import_customerHandlers.findOrCreateCustomer,
  [import_shared.FlowgladActionKey.UpdateCustomer]: import_customerHandlers.updateCustomer,
  [import_shared.FlowgladActionKey.CreateCheckoutSession]: import_checkoutSessionHandlers.createCheckoutSession,
  [import_shared.FlowgladActionKey.CreateAddPaymentMethodCheckoutSession]: import_checkoutSessionHandlers.createAddPaymentMethodCheckoutSession,
  [import_shared.FlowgladActionKey.CreateActivateSubscriptionCheckoutSession]: import_checkoutSessionHandlers.createActivateSubscriptionCheckoutSession,
  [import_shared.FlowgladActionKey.CancelSubscription]: import_subscriptionHandlers.cancelSubscription,
  [import_shared.FlowgladActionKey.UncancelSubscription]: import_subscriptionHandlers.uncancelSubscription,
  [import_shared.FlowgladActionKey.AdjustSubscription]: import_subscriptionHandlers.adjustSubscription,
  [import_shared.FlowgladActionKey.CreateSubscription]: () => __async(null, null, function* () {
    return {
      data: {},
      status: 501,
      error: {
        code: "Not Implemented",
        json: {}
      }
    };
  }),
  [import_shared.FlowgladActionKey.CreateUsageEvent]: import_usageEventHandlers.createUsageEvent,
  [import_shared.FlowgladActionKey.GetResourceUsages]: import_resourceHandlers.getResources,
  [import_shared.FlowgladActionKey.GetResourceUsage]: import_resourceHandlers.getResourceUsage,
  [import_shared.FlowgladActionKey.ClaimResource]: import_resourceHandlers.claimResource,
  [import_shared.FlowgladActionKey.ReleaseResource]: import_resourceHandlers.releaseResource,
  [import_shared.FlowgladActionKey.ListResourceClaims]: import_resourceHandlers.listResourceClaims,
  [import_shared.FlowgladActionKey.GetUsageMeterBalances]: import_usageMeterHandlers.getUsageMeterBalances
};
const hybridRouteToHandlerMap = {
  [import_shared.FlowgladActionKey.GetPricingModel]: import_pricingModelHandlers.getPricingModel
};
const isHybridActionKey = (key) => {
  return key in hybridRouteToHandlerMap;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  hybridRouteToHandlerMap,
  isHybridActionKey,
  routeToHandlerMap
});
