// @flowglad/server/dist/cjs/subrouteHandlers/pricingModelHandlers.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var pricingModelHandlers_exports = {};
__export(pricingModelHandlers_exports, {
  getPricingModel: () => getPricingModel
});
module.exports = __toCommonJS(pricingModelHandlers_exports);
const getPricingModel = (params, context) => __async(null, null, function* () {
  const { flowgladServer, flowgladServerAdmin } = context;
  if (flowgladServer) {
    try {
      const { pricingModel } = yield flowgladServer.getPricingModel();
      const response = {
        pricingModel,
        source: "customer"
      };
      return {
        data: __spreadValues({}, response),
        status: 200
      };
    } catch (error) {
      console.error(
        "[GetPricingModel] Customer pricing fetch failed after successful auth:",
        error
      );
      return {
        data: {},
        status: 500,
        error: {
          code: "PRICING_MODEL_FETCH_FAILED",
          json: {
            message: "Failed to retrieve customer pricing model",
            details: error instanceof Error ? error.message : void 0
          }
        }
      };
    }
  }
  try {
    const result = yield flowgladServerAdmin.getDefaultPricingModel();
    const pricingModel = result && typeof result === "object" && "pricingModel" in result ? result.pricingModel : result;
    const response = {
      pricingModel,
      source: "default"
    };
    return {
      data: __spreadValues({}, response),
      status: 200
    };
  } catch (error) {
    console.error(
      "[GetPricingModel] Default pricing fetch failed:",
      error
    );
    return {
      data: {},
      status: 500,
      error: {
        code: "DEFAULT_PRICING_MODEL_FETCH_FAILED",
        json: {
          message: "Failed to retrieve default pricing model",
          details: error instanceof Error ? error.message : void 0
        }
      }
    };
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getPricingModel
});
