// @flowglad/server/dist/cjs/subrouteHandlers/resourceHandlers.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var resourceHandlers_exports = {};
__export(resourceHandlers_exports, {
  claimResource: () => claimResource,
  getResourceUsage: () => getResourceUsage,
  getResources: () => getResources,
  listResourceClaims: () => listResourceClaims,
  releaseResource: () => releaseResource
});
module.exports = __toCommonJS(resourceHandlers_exports);
var import_shared = require("@flowglad/shared");
const getResources = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== import_shared.HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.getResourceUsages(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "get_resources_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const getResourceUsage = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== import_shared.HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.getResourceUsage(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "get_resource_usage_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const claimResource = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== import_shared.HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.claimResource(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "claim_resource_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const releaseResource = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== import_shared.HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.releaseResource(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "release_resource_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const listResourceClaims = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== import_shared.HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.listResourceClaims(
      params.data
    );
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "list_resource_claims_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  claimResource,
  getResourceUsage,
  getResources,
  listResourceClaims,
  releaseResource
});
