// @flowglad/server/dist/cjs/subrouteHandlers/test-utils.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var test_utils_exports = {};
__export(test_utils_exports, {
  assert200Success: () => assert200Success,
  assert400BadRequest: () => assert400BadRequest,
  assert401Unauthorized: () => assert401Unauthorized,
  assert403Forbidden: () => assert403Forbidden,
  assert404NotFound: () => assert404NotFound,
  assert405MethodNotAllowed: () => assert405MethodNotAllowed,
  assert500Error: () => assert500Error,
  assertHandlerResponse: () => assertHandlerResponse,
  jsonMessageContains: () => jsonMessageContains,
  jsonMessageMatches: () => jsonMessageMatches
});
module.exports = __toCommonJS(test_utils_exports);
var import_vitest = require("vitest");
const assertHandlerResponse = (actual, expected) => {
  (0, import_vitest.expect)(actual.status).toBe(expected.status);
  if (expected.error === "undefined") {
    (0, import_vitest.expect)(actual.error).toBeUndefined();
  } else if (expected.error !== void 0) {
    (0, import_vitest.expect)(actual.error).not.toBeUndefined();
    (0, import_vitest.expect)(actual.error.code).toBe(expected.error.code);
    if (expected.error.json !== void 0) {
      if (typeof expected.error.json === "function") {
        expected.error.json(actual.error.json);
      } else {
        (0, import_vitest.expect)(actual.error.json).toEqual(expected.error.json);
      }
    }
  }
  if (expected.data !== void 0) {
    if (typeof expected.data === "function") {
      expected.data(actual.data);
    } else {
      (0, import_vitest.expect)(actual.data).toEqual(expected.data);
    }
  }
};
const assert405MethodNotAllowed = (actual, variant = "standard") => {
  assertHandlerResponse(actual, {
    status: 405,
    error: variant === "standard" ? { code: "Method not allowed", json: {} } : { code: "405", json: { message: "Method not allowed" } },
    data: {}
  });
};
const assert200Success = (actual, expectedData) => {
  assertHandlerResponse(actual, {
    status: 200,
    error: "undefined",
    data: expectedData
  });
};
const assert500Error = (actual, errorCode, message) => {
  assertHandlerResponse(actual, {
    status: 500,
    error: {
      code: errorCode,
      json: { message }
    },
    data: {}
  });
};
const assert401Unauthorized = (actual) => {
  assertHandlerResponse(actual, {
    status: 401,
    error: {
      code: "401",
      json: { message: "Unauthorized" }
    },
    data: {}
  });
};
const assert403Forbidden = (actual, message) => {
  assertHandlerResponse(actual, {
    status: 403,
    error: {
      code: "forbidden",
      json: { message }
    },
    data: {}
  });
};
const assert404NotFound = (actual, message) => {
  assertHandlerResponse(actual, {
    status: 404,
    error: {
      code: "404",
      json: { message }
    },
    data: {}
  });
};
const assert400BadRequest = (actual, errorCode, message) => {
  assertHandlerResponse(actual, {
    status: 400,
    error: {
      code: errorCode,
      json: { message }
    },
    data: {}
  });
};
const jsonMessageContains = (substring) => {
  return (json) => {
    (0, import_vitest.expect)(json.message).not.toBeUndefined();
    (0, import_vitest.expect)(String(json.message)).toContain(substring);
  };
};
const jsonMessageMatches = (pattern) => {
  return (json) => {
    (0, import_vitest.expect)(json.message).not.toBeUndefined();
    (0, import_vitest.expect)(String(json.message)).toMatch(pattern);
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  assert200Success,
  assert400BadRequest,
  assert401Unauthorized,
  assert403Forbidden,
  assert404NotFound,
  assert405MethodNotAllowed,
  assert500Error,
  assertHandlerResponse,
  jsonMessageContains,
  jsonMessageMatches
});
