// @flowglad/server/dist/cjs/subrouteHandlers/usageEventHandlers.cjs
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var usageEventHandlers_exports = {};
__export(usageEventHandlers_exports, {
  createUsageEvent: () => createUsageEvent
});
module.exports = __toCommonJS(usageEventHandlers_exports);
var import_shared = require("@flowglad/shared");
var import_nanoid = require("nanoid");
var import_serverUtils = require('../serverUtils.cjs');
const createUsageEvent = (params, flowgladServer) => __async(null, null, function* () {
  var _a, _b, _c, _d, _e;
  let error;
  let status;
  let data = {};
  if (params.method !== import_shared.HTTPMethod.POST) {
    error = {
      code: "Method not allowed",
      json: {}
    };
    status = 405;
    return {
      data,
      status,
      error
    };
  }
  try {
    const billing = yield flowgladServer.getBilling();
    const currentSubscription = billing.currentSubscription;
    const subscriptionId = (_a = params.data.subscriptionId) != null ? _a : currentSubscription == null ? void 0 : currentSubscription.id;
    if (!subscriptionId) {
      return {
        data: {},
        status: 400,
        error: {
          code: "missing_subscription_id",
          json: {
            message: "subscriptionId required: no current subscription found"
          }
        }
      };
    }
    const customerSubscriptionIds = (_c = (_b = billing.currentSubscriptions) == null ? void 0 : _b.map((sub) => sub.id)) != null ? _c : [];
    if (!customerSubscriptionIds.includes(subscriptionId)) {
      return {
        data: {},
        status: 403,
        error: {
          code: "forbidden",
          json: {
            message: `Subscription ${subscriptionId} is not found among the customer's current subscriptions`
          }
        }
      };
    }
    const resolvedParams = __spreadProps(__spreadValues({}, params.data), {
      subscriptionId,
      amount: (_d = params.data.amount) != null ? _d : 1,
      transactionId: (_e = params.data.transactionId) != null ? _e : (0, import_nanoid.nanoid)()
    });
    const usageEvent = yield flowgladServer.createUsageEvent(resolvedParams);
    data = usageEvent;
    status = 200;
  } catch (e) {
    if (e instanceof Error) {
      error = (0, import_serverUtils.parseErrorStringToErrorObject)(e.message);
    } else {
      error = {
        code: "Unknown error",
        json: {}
      };
    }
    status = 500;
  }
  return {
    data,
    status,
    error
  };
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createUsageEvent
});
