// @flowglad/react/dist/cjs/FlowgladConfigContext.js
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FlowgladConfigContext_exports = {};
__export(FlowgladConfigContext_exports, {
  FlowgladConfigProvider: () => FlowgladConfigProvider,
  useFlowgladConfig: () => useFlowgladConfig
});
module.exports = __toCommonJS(FlowgladConfigContext_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
const defaultConfigContextValues = {
  baseURL: void 0,
  betterAuthBasePath: void 0,
  requestConfig: void 0,
  __devMode: false,
  billingMocks: void 0
};
const FlowgladConfigContext = (0, import_react.createContext)(
  defaultConfigContextValues
);
const FlowgladConfigProvider = ({
  children,
  baseURL,
  betterAuthBasePath,
  requestConfig,
  __devMode = false,
  billingMocks
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    FlowgladConfigContext.Provider,
    {
      value: {
        baseURL,
        betterAuthBasePath,
        requestConfig,
        __devMode,
        billingMocks
      },
      children
    }
  );
};
const useFlowgladConfig = () => {
  return (0, import_react.useContext)(FlowgladConfigContext);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowgladConfigProvider,
  useFlowgladConfig
});
