// @flowglad/react/dist/cjs/useUsageMeters.js
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var useUsageMeters_exports = {};
__export(useUsageMeters_exports, {
  USAGE_METERS_QUERY_KEY: () => USAGE_METERS_QUERY_KEY,
  useUsageMeter: () => useUsageMeter,
  useUsageMeters: () => useUsageMeters
});
module.exports = __toCommonJS(useUsageMeters_exports);
var import_shared = require("@flowglad/shared");
var import_react_query = require("@tanstack/react-query");
var import_FlowgladConfigContext = require("./FlowgladConfigContext");
var import_FlowgladContext = require("./FlowgladContext");
const USAGE_METERS_QUERY_KEY = "flowglad-usage-meters";
const deriveUsageMeterBalancesFromBillingMocks = (billingMocks) => {
  var _a;
  const currentSubscriptions = (_a = billingMocks.currentSubscriptions) != null ? _a : [];
  const balances = [];
  for (const subscription of currentSubscriptions) {
    const experimental = subscription.experimental;
    if (experimental == null ? void 0 : experimental.usageMeterBalances) {
      balances.push(...experimental.usageMeterBalances);
    }
  }
  return balances;
};
const useUsageMeters = (params) => {
  var _a, _b, _c, _d, _e;
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    __devMode,
    billingMocks
  } = (0, import_FlowgladConfigContext.useFlowgladConfig)();
  const {
    data: responseData,
    isLoading,
    error
  } = (0, import_react_query.useQuery)({
    queryKey: [USAGE_METERS_QUERY_KEY, params == null ? void 0 : params.subscriptionId],
    enabled: !__devMode,
    queryFn: () => __async(null, null, function* () {
      const flowgladRoute = (0, import_FlowgladContext.getFlowgladRoute)(
        baseURL,
        betterAuthBasePath
      );
      const response = yield fetch(
        `${flowgladRoute}/${import_shared.FlowgladActionKey.GetUsageMeterBalances}`,
        {
          method: import_shared.flowgladActionValidators[import_shared.FlowgladActionKey.GetUsageMeterBalances].method,
          headers: __spreadValues({
            "Content-Type": "application/json"
          }, requestConfig == null ? void 0 : requestConfig.headers),
          body: JSON.stringify(params != null ? params : {})
        }
      );
      const json = yield response.json();
      return json;
    })
  });
  if (__devMode) {
    if (!billingMocks) {
      throw new Error(
        "FlowgladProvider: __devMode requires billingMocks"
      );
    }
    const balances = deriveUsageMeterBalancesFromBillingMocks(billingMocks);
    const filteredBalances = (params == null ? void 0 : params.subscriptionId) ? balances.filter(
      (b) => b.subscriptionId === params.subscriptionId
    ) : balances;
    return {
      usageMeters: filteredBalances,
      isLoading: false,
      error: null
    };
  }
  if (responseData == null ? void 0 : responseData.error) {
    return {
      usageMeters: void 0,
      isLoading: false,
      error: new Error(
        (_d = (_c = (_b = (_a = responseData.error.json) == null ? void 0 : _a.message) == null ? void 0 : _b.toString()) != null ? _c : responseData.error.code) != null ? _d : "Failed to fetch usage meter balances"
      )
    };
  }
  return {
    usageMeters: (_e = responseData == null ? void 0 : responseData.data) == null ? void 0 : _e.usageMeterBalances,
    isLoading,
    error: error != null ? error : null
  };
};
const useUsageMeter = (usageMeterSlug, params) => {
  var _a;
  const { usageMeters, isLoading, error } = useUsageMeters(params);
  const usageMeter = (_a = usageMeters == null ? void 0 : usageMeters.find((b) => b.slug === usageMeterSlug)) != null ? _a : null;
  return { usageMeter, isLoading, error };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  USAGE_METERS_QUERY_KEY,
  useUsageMeter,
  useUsageMeters
});
