// @flowglad/shared/dist/cjs/types/sdk.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sdk_exports = {};
__export(sdk_exports, {
  FlowgladActionKey: () => FlowgladActionKey,
  HTTPMethod: () => HTTPMethod
});
module.exports = __toCommonJS(sdk_exports);
var FlowgladActionKey = /* @__PURE__ */ ((FlowgladActionKey2) => {
  FlowgladActionKey2["GetCustomerBilling"] = "customers/billing";
  FlowgladActionKey2["FindOrCreateCustomer"] = "customers/find-or-create";
  FlowgladActionKey2["CreateCheckoutSession"] = "checkout-sessions/create";
  FlowgladActionKey2["CreateAddPaymentMethodCheckoutSession"] = "checkout-sessions/create-add-payment-method";
  FlowgladActionKey2["CreateActivateSubscriptionCheckoutSession"] = "checkout-sessions/create-activate-subscription";
  FlowgladActionKey2["CancelSubscription"] = "subscriptions/cancel";
  FlowgladActionKey2["UncancelSubscription"] = "subscriptions/uncancel";
  FlowgladActionKey2["AdjustSubscription"] = "subscriptions/adjust";
  FlowgladActionKey2["CreateSubscription"] = "subscriptions/create";
  FlowgladActionKey2["UpdateCustomer"] = "customers/update";
  FlowgladActionKey2["CreateUsageEvent"] = "usage-events/create";
  FlowgladActionKey2["GetResourceUsages"] = "resources/usages";
  FlowgladActionKey2["GetResourceUsage"] = "resources/usages/retrieve";
  FlowgladActionKey2["ClaimResource"] = "resources/claim";
  FlowgladActionKey2["ReleaseResource"] = "resources/release";
  FlowgladActionKey2["ListResourceClaims"] = "resources/claims";
  FlowgladActionKey2["GetPricingModel"] = "pricing-models/retrieve";
  FlowgladActionKey2["GetUsageMeterBalances"] = "usage-meters/balances";
  return FlowgladActionKey2;
})(FlowgladActionKey || {});
var HTTPMethod = /* @__PURE__ */ ((HTTPMethod2) => {
  HTTPMethod2["GET"] = "GET";
  HTTPMethod2["POST"] = "POST";
  HTTPMethod2["PUT"] = "PUT";
  HTTPMethod2["PATCH"] = "PATCH";
  HTTPMethod2["DELETE"] = "DELETE";
  return HTTPMethod2;
})(HTTPMethod || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowgladActionKey,
  HTTPMethod
});
