// @flowglad/shared/dist/esm/types/sdk.mjs
import "../chunk-2MM5EJJX.mjs";
var FlowgladActionKey = /* @__PURE__ */ ((FlowgladActionKey2) => {
  FlowgladActionKey2["GetCustomerBilling"] = "customers/billing";
  FlowgladActionKey2["FindOrCreateCustomer"] = "customers/find-or-create";
  FlowgladActionKey2["CreateCheckoutSession"] = "checkout-sessions/create";
  FlowgladActionKey2["CreateAddPaymentMethodCheckoutSession"] = "checkout-sessions/create-add-payment-method";
  FlowgladActionKey2["CreateActivateSubscriptionCheckoutSession"] = "checkout-sessions/create-activate-subscription";
  FlowgladActionKey2["CancelSubscription"] = "subscriptions/cancel";
  FlowgladActionKey2["UncancelSubscription"] = "subscriptions/uncancel";
  FlowgladActionKey2["AdjustSubscription"] = "subscriptions/adjust";
  FlowgladActionKey2["CreateSubscription"] = "subscriptions/create";
  FlowgladActionKey2["UpdateCustomer"] = "customers/update";
  FlowgladActionKey2["CreateUsageEvent"] = "usage-events/create";
  FlowgladActionKey2["GetResourceUsages"] = "resources/usages";
  FlowgladActionKey2["GetResourceUsage"] = "resources/usages/retrieve";
  FlowgladActionKey2["ClaimResource"] = "resources/claim";
  FlowgladActionKey2["ReleaseResource"] = "resources/release";
  FlowgladActionKey2["ListResourceClaims"] = "resources/claims";
  FlowgladActionKey2["GetPricingModel"] = "pricing-models/retrieve";
  FlowgladActionKey2["GetUsageMeterBalances"] = "usage-meters/balances";
  return FlowgladActionKey2;
})(FlowgladActionKey || {});
var HTTPMethod = /* @__PURE__ */ ((HTTPMethod2) => {
  HTTPMethod2["GET"] = "GET";
  HTTPMethod2["POST"] = "POST";
  HTTPMethod2["PUT"] = "PUT";
  HTTPMethod2["PATCH"] = "PATCH";
  HTTPMethod2["DELETE"] = "DELETE";
  return HTTPMethod2;
})(HTTPMethod || {});
export {
  FlowgladActionKey,
  HTTPMethod
};
