// @flowglad/node/resources/catalogs.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Catalogs = void 0;
const resource_1 = require("../core/resource.js");
const path_1 = require("../internal/utils/path.js");
class Catalogs extends resource_1.APIResource {
    /**
     * Create Catalog
     */
    create(body, options) {
        return this._client.post('/api/v1/catalogs', { body, ...options });
    }
    /**
     * Get Catalog
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/api/v1/catalogs/${id}`, options);
    }
    /**
     * Update Catalog
     */
    update(id, body, options) {
        return this._client.put((0, path_1.path) `/api/v1/catalogs/${id}`, { body, ...options });
    }
    /**
     * List Catalogs
     */
    list(query = {}, options) {
        return this._client.get('/api/v1/catalogs', { query, ...options });
    }
    /**
     * Clone a Catalog
     */
    clone(id, body, options) {
        return this._client.post((0, path_1.path) `/api/v1/catalogs/${id}/clone`, { body, ...options });
    }
    /**
     * Get Default Catalog for Organization
     */
    retrieveDefault(options) {
        return this._client.get('/api/v1/catalogs/default', options);
    }
}
exports.Catalogs = Catalogs;
