// @flowglad/node/resources/catalogs.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { path } from "../internal/utils/path.mjs";
export class Catalogs extends APIResource {
    /**
     * Create Catalog
     */
    create(body, options) {
        return this._client.post('/api/v1/catalogs', { body, ...options });
    }
    /**
     * Get Catalog
     */
    retrieve(id, options) {
        return this._client.get(path `/api/v1/catalogs/${id}`, options);
    }
    /**
     * Update Catalog
     */
    update(id, body, options) {
        return this._client.put(path `/api/v1/catalogs/${id}`, { body, ...options });
    }
    /**
     * List Catalogs
     */
    list(query = {}, options) {
        return this._client.get('/api/v1/catalogs', { query, ...options });
    }
    /**
     * Clone a Catalog
     */
    clone(id, body, options) {
        return this._client.post(path `/api/v1/catalogs/${id}/clone`, { body, ...options });
    }
    /**
     * Get Default Catalog for Organization
     */
    retrieveDefault(options) {
        return this._client.get('/api/v1/catalogs/default', options);
    }
}
