// @flowglad/node/resources/customers.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { path } from "../internal/utils/path.mjs";
export class Customers extends APIResource {
    /**
     * Create Customer
     */
    create(body, options) {
        return this._client.post('/api/v1/customers', { body, ...options });
    }
    /**
     * Get Customer
     */
    retrieve(externalID, options) {
        return this._client.get(path `/api/v1/customers/${externalID}`, options);
    }
    /**
     * Update Customer
     */
    update(externalID, body, options) {
        return this._client.put(path `/api/v1/customers/${externalID}`, { body, ...options });
    }
    /**
     * List Customers
     */
    list(query = {}, options) {
        return this._client.get('/api/v1/customers', { query, ...options });
    }
    /**
     * Get Billing Details
     */
    retrieveBilling(externalID, options) {
        return this._client.get(path `/api/v1/customers/${externalID}/billing`, options);
    }
}
