// @flowglad/node/resources/customers.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Customers = void 0;
const resource_1 = require("../core/resource.js");
const path_1 = require("../internal/utils/path.js");
class Customers extends resource_1.APIResource {
    /**
     * Create Customer
     */
    create(body, options) {
        return this._client.post('/api/v1/customers', { body, ...options });
    }
    /**
     * Get Customer
     */
    retrieve(externalID, options) {
        return this._client.get((0, path_1.path) `/api/v1/customers/${externalID}`, options);
    }
    /**
     * Update Customer
     */
    update(externalID, body, options) {
        return this._client.put((0, path_1.path) `/api/v1/customers/${externalID}`, { body, ...options });
    }
    /**
     * List Customers
     */
    list(query = {}, options) {
        return this._client.get('/api/v1/customers', { query, ...options });
    }
    /**
     * Get Billing Details
     */
    retrieveBilling(externalID, options) {
        return this._client.get((0, path_1.path) `/api/v1/customers/${externalID}/billing`, options);
    }
}
exports.Customers = Customers;
