// @flowglad/node/resources/pricing-models.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { path } from "../internal/utils/path.mjs";
export class PricingModels extends APIResource {
    /**
     * Create Pricing Model
     */
    create(body, options) {
        return this._client.post('/api/v1/pricing-models', { body, ...options });
    }
    /**
     * Get Pricing Model
     */
    retrieve(id, options) {
        return this._client.get(path `/api/v1/pricing-models/${id}`, options);
    }
    /**
     * Update Pricing Model
     */
    update(id, body, options) {
        return this._client.put(path `/api/v1/pricing-models/${id}`, { body, ...options });
    }
    /**
     * List Pricing Models
     */
    list(query = {}, options) {
        return this._client.get('/api/v1/pricing-models', { query, ...options });
    }
    /**
     * Clone a PricingModel
     */
    clone(id, body, options) {
        return this._client.post(path `/api/v1/pricing-models/${id}/clone`, { body, ...options });
    }
    /**
     * Get Default Pricing Model for Organization
     */
    retrieveDefault(options) {
        return this._client.get('/api/v1/pricing-models/default', options);
    }
}
