// @flowglad/server/dist/esm/subrouteHandlers/index.mjs
import {
  __async
} from "../chunk-E3LOUS7X.mjs";
import {
  FlowgladActionKey
} from "@flowglad/shared";
import {
  createActivateSubscriptionCheckoutSession,
  createAddPaymentMethodCheckoutSession,
  createCheckoutSession
} from "./checkoutSessionHandlers.mjs";
import {
  findOrCreateCustomer,
  getCustomerBilling,
  updateCustomer
} from "./customerHandlers.mjs";
import { getPricingModel } from "./pricingModelHandlers.mjs";
import {
  claimResource,
  getResources,
  getResourceUsage,
  listResourceClaims,
  releaseResource
} from "./resourceHandlers.mjs";
import {
  adjustSubscription,
  cancelSubscription,
  uncancelSubscription
} from "./subscriptionHandlers.mjs";
import { createUsageEvent } from "./usageEventHandlers.mjs";
import { getUsageMeterBalances } from "./usageMeterHandlers.mjs";
const routeToHandlerMap = {
  [FlowgladActionKey.GetCustomerBilling]: getCustomerBilling,
  [FlowgladActionKey.FindOrCreateCustomer]: findOrCreateCustomer,
  [FlowgladActionKey.UpdateCustomer]: updateCustomer,
  [FlowgladActionKey.CreateCheckoutSession]: createCheckoutSession,
  [FlowgladActionKey.CreateAddPaymentMethodCheckoutSession]: createAddPaymentMethodCheckoutSession,
  [FlowgladActionKey.CreateActivateSubscriptionCheckoutSession]: createActivateSubscriptionCheckoutSession,
  [FlowgladActionKey.CancelSubscription]: cancelSubscription,
  [FlowgladActionKey.UncancelSubscription]: uncancelSubscription,
  [FlowgladActionKey.AdjustSubscription]: adjustSubscription,
  [FlowgladActionKey.CreateSubscription]: () => __async(null, null, function* () {
    return {
      data: {},
      status: 501,
      error: {
        code: "Not Implemented",
        json: {}
      }
    };
  }),
  [FlowgladActionKey.CreateUsageEvent]: createUsageEvent,
  [FlowgladActionKey.GetResourceUsages]: getResources,
  [FlowgladActionKey.GetResourceUsage]: getResourceUsage,
  [FlowgladActionKey.ClaimResource]: claimResource,
  [FlowgladActionKey.ReleaseResource]: releaseResource,
  [FlowgladActionKey.ListResourceClaims]: listResourceClaims,
  [FlowgladActionKey.GetUsageMeterBalances]: getUsageMeterBalances
};
const hybridRouteToHandlerMap = {
  [FlowgladActionKey.GetPricingModel]: getPricingModel
};
const isHybridActionKey = (key) => {
  return key in hybridRouteToHandlerMap;
};
export {
  hybridRouteToHandlerMap,
  isHybridActionKey,
  routeToHandlerMap
};
