// @flowglad/server/dist/esm/subrouteHandlers/pricingModelHandlers.mjs
import {
  __async,
  __spreadValues
} from "../chunk-E3LOUS7X.mjs";
const getPricingModel = (params, context) => __async(null, null, function* () {
  const { flowgladServer, flowgladServerAdmin } = context;
  if (flowgladServer) {
    try {
      const { pricingModel } = yield flowgladServer.getPricingModel();
      const response = {
        pricingModel,
        source: "customer"
      };
      return {
        data: __spreadValues({}, response),
        status: 200
      };
    } catch (error) {
      console.error(
        "[GetPricingModel] Customer pricing fetch failed after successful auth:",
        error
      );
      return {
        data: {},
        status: 500,
        error: {
          code: "PRICING_MODEL_FETCH_FAILED",
          json: {
            message: "Failed to retrieve customer pricing model",
            details: error instanceof Error ? error.message : void 0
          }
        }
      };
    }
  }
  try {
    const result = yield flowgladServerAdmin.getDefaultPricingModel();
    const pricingModel = result && typeof result === "object" && "pricingModel" in result ? result.pricingModel : result;
    const response = {
      pricingModel,
      source: "default"
    };
    return {
      data: __spreadValues({}, response),
      status: 200
    };
  } catch (error) {
    console.error(
      "[GetPricingModel] Default pricing fetch failed:",
      error
    );
    return {
      data: {},
      status: 500,
      error: {
        code: "DEFAULT_PRICING_MODEL_FETCH_FAILED",
        json: {
          message: "Failed to retrieve default pricing model",
          details: error instanceof Error ? error.message : void 0
        }
      }
    };
  }
});
export {
  getPricingModel
};
