// @flowglad/server/dist/esm/subrouteHandlers/resourceHandlers.mjs
import {
  __async
} from "../chunk-E3LOUS7X.mjs";
import { HTTPMethod } from "@flowglad/shared";
const getResources = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.getResourceUsages(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "get_resources_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const getResourceUsage = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.getResourceUsage(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "get_resource_usage_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const claimResource = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.claimResource(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "claim_resource_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const releaseResource = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.releaseResource(params.data);
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "release_resource_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
const listResourceClaims = (params, flowgladServer) => __async(null, null, function* () {
  if (params.method !== HTTPMethod.POST) {
    return {
      data: {},
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.listResourceClaims(
      params.data
    );
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "list_resource_claims_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
export {
  claimResource,
  getResourceUsage,
  getResources,
  listResourceClaims,
  releaseResource
};
