// @flowglad/server/dist/esm/subrouteHandlers/subscriptionHandlers.mjs
import {
  __async
} from "../chunk-E3LOUS7X.mjs";
import { HTTPMethod } from "@flowglad/shared";
const cancelSubscription = (params, flowgladServer) => __async(null, null, function* () {
  let error;
  let status;
  const data = {};
  if (params.method !== HTTPMethod.POST) {
    error = {
      code: "Method not allowed",
      json: {}
    };
    status = 405;
    return {
      data,
      status,
      error
    };
  }
  let subscription;
  try {
    subscription = yield flowgladServer.cancelSubscription(
      params.data
    );
    return {
      data: subscription,
      status: 200
    };
  } catch (error2) {
    return {
      data: {},
      status: 500,
      error: {
        code: "subscription_cancel_failed",
        json: {
          message: error2.message
        }
      }
    };
  }
});
const uncancelSubscription = (params, flowgladServer) => __async(null, null, function* () {
  let error;
  let status;
  const data = {};
  if (params.method !== HTTPMethod.POST) {
    error = {
      code: "Method not allowed",
      json: {}
    };
    status = 405;
    return {
      data,
      status,
      error
    };
  }
  let subscription;
  try {
    subscription = yield flowgladServer.uncancelSubscription(
      params.data
    );
    return {
      data: subscription,
      status: 200
    };
  } catch (error2) {
    return {
      data: {},
      status: 500,
      error: {
        code: "subscription_uncancel_failed",
        json: {
          message: error2.message
        }
      }
    };
  }
});
const adjustSubscription = (params, flowgladServer) => __async(null, null, function* () {
  const data = {};
  if (params.method !== HTTPMethod.POST) {
    return {
      data,
      status: 405,
      error: {
        code: "Method not allowed",
        json: {}
      }
    };
  }
  try {
    const result = yield flowgladServer.adjustSubscription(
      params.data
    );
    return {
      data: result,
      status: 200
    };
  } catch (error) {
    return {
      data: {},
      status: 500,
      error: {
        code: "subscription_adjust_failed",
        json: {
          message: error.message
        }
      }
    };
  }
});
export {
  adjustSubscription,
  cancelSubscription,
  uncancelSubscription
};
