// @flowglad/server/dist/esm/subrouteHandlers/test-utils.mjs
import "../chunk-E3LOUS7X.mjs";
import { expect } from "vitest";
const assertHandlerResponse = (actual, expected) => {
  expect(actual.status).toBe(expected.status);
  if (expected.error === "undefined") {
    expect(actual.error).toBeUndefined();
  } else if (expected.error !== void 0) {
    expect(actual.error).not.toBeUndefined();
    expect(actual.error.code).toBe(expected.error.code);
    if (expected.error.json !== void 0) {
      if (typeof expected.error.json === "function") {
        expected.error.json(actual.error.json);
      } else {
        expect(actual.error.json).toEqual(expected.error.json);
      }
    }
  }
  if (expected.data !== void 0) {
    if (typeof expected.data === "function") {
      expected.data(actual.data);
    } else {
      expect(actual.data).toEqual(expected.data);
    }
  }
};
const assert405MethodNotAllowed = (actual, variant = "standard") => {
  assertHandlerResponse(actual, {
    status: 405,
    error: variant === "standard" ? { code: "Method not allowed", json: {} } : { code: "405", json: { message: "Method not allowed" } },
    data: {}
  });
};
const assert200Success = (actual, expectedData) => {
  assertHandlerResponse(actual, {
    status: 200,
    error: "undefined",
    data: expectedData
  });
};
const assert500Error = (actual, errorCode, message) => {
  assertHandlerResponse(actual, {
    status: 500,
    error: {
      code: errorCode,
      json: { message }
    },
    data: {}
  });
};
const assert401Unauthorized = (actual) => {
  assertHandlerResponse(actual, {
    status: 401,
    error: {
      code: "401",
      json: { message: "Unauthorized" }
    },
    data: {}
  });
};
const assert403Forbidden = (actual, message) => {
  assertHandlerResponse(actual, {
    status: 403,
    error: {
      code: "forbidden",
      json: { message }
    },
    data: {}
  });
};
const assert404NotFound = (actual, message) => {
  assertHandlerResponse(actual, {
    status: 404,
    error: {
      code: "404",
      json: { message }
    },
    data: {}
  });
};
const assert400BadRequest = (actual, errorCode, message) => {
  assertHandlerResponse(actual, {
    status: 400,
    error: {
      code: errorCode,
      json: { message }
    },
    data: {}
  });
};
const jsonMessageContains = (substring) => {
  return (json) => {
    expect(json.message).not.toBeUndefined();
    expect(String(json.message)).toContain(substring);
  };
};
const jsonMessageMatches = (pattern) => {
  return (json) => {
    expect(json.message).not.toBeUndefined();
    expect(String(json.message)).toMatch(pattern);
  };
};
export {
  assert200Success,
  assert400BadRequest,
  assert401Unauthorized,
  assert403Forbidden,
  assert404NotFound,
  assert405MethodNotAllowed,
  assert500Error,
  assertHandlerResponse,
  jsonMessageContains,
  jsonMessageMatches
};
