// @flowglad/server/dist/esm/subrouteHandlers/usageEventHandlers.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "../chunk-E3LOUS7X.mjs";
import { HTTPMethod } from "@flowglad/shared";
import { nanoid } from "nanoid";
import { parseErrorStringToErrorObject } from "../serverUtils.mjs";
const createUsageEvent = (params, flowgladServer) => __async(null, null, function* () {
  var _a, _b, _c, _d, _e;
  let error;
  let status;
  let data = {};
  if (params.method !== HTTPMethod.POST) {
    error = {
      code: "Method not allowed",
      json: {}
    };
    status = 405;
    return {
      data,
      status,
      error
    };
  }
  try {
    const billing = yield flowgladServer.getBilling();
    const currentSubscription = billing.currentSubscription;
    const subscriptionId = (_a = params.data.subscriptionId) != null ? _a : currentSubscription == null ? void 0 : currentSubscription.id;
    if (!subscriptionId) {
      return {
        data: {},
        status: 400,
        error: {
          code: "missing_subscription_id",
          json: {
            message: "subscriptionId required: no current subscription found"
          }
        }
      };
    }
    const customerSubscriptionIds = (_c = (_b = billing.currentSubscriptions) == null ? void 0 : _b.map((sub) => sub.id)) != null ? _c : [];
    if (!customerSubscriptionIds.includes(subscriptionId)) {
      return {
        data: {},
        status: 403,
        error: {
          code: "forbidden",
          json: {
            message: `Subscription ${subscriptionId} is not found among the customer's current subscriptions`
          }
        }
      };
    }
    const resolvedParams = __spreadProps(__spreadValues({}, params.data), {
      subscriptionId,
      amount: (_d = params.data.amount) != null ? _d : 1,
      transactionId: (_e = params.data.transactionId) != null ? _e : nanoid()
    });
    const usageEvent = yield flowgladServer.createUsageEvent(resolvedParams);
    data = usageEvent;
    status = 200;
  } catch (e) {
    if (e instanceof Error) {
      error = parseErrorStringToErrorObject(e.message);
    } else {
      error = {
        code: "Unknown error",
        json: {}
      };
    }
    status = 500;
  }
  return {
    data,
    status,
    error
  };
});
export {
  createUsageEvent
};
