// @flowglad/server/dist/esm/supabase/supabaseEdgeHandler.mjs
import {
  __async
} from "../chunk-E3LOUS7X.mjs";
import {
  requestHandler
} from "../requestHandler.mjs";
const extractPath = (pathname) => {
  const trimmedPath = pathname.replace(/^\/+|\/+$/g, "");
  return trimmedPath === "" ? [] : trimmedPath.split("/");
};
const supabaseEdgeHandler = (options) => {
  const handler = requestHandler(options);
  return (req) => __async(null, null, function* () {
    var _a, _b;
    let url;
    try {
      url = new URL(req.url);
    } catch (err) {
      if (options.onError) {
        options.onError(err);
      }
      return new Response(
        JSON.stringify({
          data: null,
          error: { message: "Invalid request URL" }
        }),
        {
          status: 400,
          headers: { "Content-Type": "application/json" }
        }
      );
    }
    let pathSegments = [];
    if (options.basePath) {
      const basePathTrimmed = options.basePath.replace(
        /^\/+|\/+$/g,
        ""
      );
      const basePathWithSlash = `/${basePathTrimmed}`;
      if (url.pathname.startsWith(basePathWithSlash)) {
        const relativePath = url.pathname.slice(
          basePathWithSlash.length
        );
        pathSegments = extractPath(relativePath);
      } else {
        pathSegments = extractPath(url.pathname);
      }
    } else {
      const segments = extractPath(url.pathname);
      if (segments.length >= 3 && segments[0] === "functions" && segments[1] === "v1") {
        pathSegments = segments.slice(3);
      } else {
        pathSegments = segments;
      }
    }
    const queryParams = Array.from(url.searchParams.entries());
    const query = queryParams.length > 0 ? Object.fromEntries(queryParams) : void 0;
    let body = void 0;
    if (req.method !== "GET") {
      try {
        body = yield req.json();
      } catch (e) {
      }
    }
    try {
      const result = yield handler(
        {
          path: pathSegments,
          method: req.method,
          query,
          body
        },
        req
      );
      return new Response(
        JSON.stringify({
          data: (_a = result.data) != null ? _a : null,
          error: (_b = result.error) != null ? _b : null
        }),
        {
          status: result.status,
          headers: {
            "Content-Type": "application/json"
          }
        }
      );
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : "Internal server error";
      return new Response(
        JSON.stringify({
          data: null,
          error: { message: errorMessage }
        }),
        {
          status: 500,
          headers: { "Content-Type": "application/json" }
        }
      );
    }
  });
};
export {
  supabaseEdgeHandler
};
