// @flowglad/server/dist/cjs/express/expressRouter.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var expressRouter_exports = {};
__export(expressRouter_exports, {
  expressRouter: () => expressRouter
});
module.exports = __toCommonJS(expressRouter_exports);
var import_express = require("express");
var import_requestHandler = require('../requestHandler.cjs');
const normalizeQueryParameters = (query) => {
  return Object.fromEntries(
    Object.entries(query).map(([key, value]) => [
      key,
      Array.isArray(value) ? value[0] : typeof value === "object" && value !== null ? null : String(value)
    ]).filter(
      (entry) => entry[1] !== null && entry[1] !== void 0 && entry[1] !== "undefined"
    )
  );
};
const expressRouter = (options) => {
  const router = (0, import_express.Router)();
  const handler = (0, import_requestHandler.requestHandler)(options);
  if (options.middleware) {
    router.use(...options.middleware);
  }
  router.all("*", (req, res) => __async(null, null, function* () {
    const pathFragments = req.path.split("/").filter((fragment) => fragment !== "");
    const result = yield handler(
      {
        path: pathFragments,
        method: req.method,
        query: req.method === "GET" ? normalizeQueryParameters(req.query) : void 0,
        body: req.method !== "GET" ? req.body : void 0
      },
      req
    );
    res.status(result.status).json({
      error: result.error,
      data: result.data
    });
  }));
  return router;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  expressRouter
});
