// @flowglad/react/dist/esm/lib/utils.js
import "../chunk-CRTRMMJ7.js";
import { clsx } from "clsx";
import { format } from "date-fns";
import { twMerge } from "tailwind-merge";
function cn(...inputs) {
  return twMerge(clsx(inputs));
}
function currencyCharacter(currency, locale) {
  const defaultLocale = typeof navigator !== "undefined" ? navigator.language : "en-US";
  return 0 .toLocaleString(locale != null ? locale : defaultLocale, {
    style: "currency",
    currency,
    minimumFractionDigits: 0,
    maximumFractionDigits: 0
  }).replace(/\d/g, "").trim();
}
function humanReadableCurrencyAmount(currency, amount) {
  const formatter = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency
  });
  if (!isCurrencyZeroDecimal(currency)) {
    return formatter.format(Number((amount / 100).toFixed(2)));
  }
  return formatter.format(amount);
}
const zeroDecimalCurrencies = [
  "BIF",
  "CLP",
  "DJF",
  "GNF",
  "JPY",
  "KMF",
  "KRW",
  "MGA",
  "PYG",
  "RWF",
  "UGX",
  "VND",
  "VUV",
  "XAF",
  "XOF",
  "XPF"
];
function isCurrencyZeroDecimal(currency) {
  return zeroDecimalCurrencies.includes(currency);
}
function formatDate(date) {
  return format(
    date instanceof Date ? date : new Date(date),
    "MMM d, yyyy"
  );
}
function devWarn(message) {
  if (process.env.NODE_ENV !== "production") {
    console.warn(`[flowglad]: ${message}`);
  }
}
function devError(message) {
  if (process.env.NODE_ENV !== "production") {
    console.error(`[flowglad]: ${message}`);
  }
}
export {
  cn,
  currencyCharacter,
  devError,
  devWarn,
  formatDate,
  humanReadableCurrencyAmount,
  isCurrencyZeroDecimal,
  zeroDecimalCurrencies
};
