// @flowglad/server/dist/cjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  FlowgladServer: () => import_FlowgladServer.FlowgladServer,
  FlowgladServerAdmin: () => import_FlowgladServerAdmin.FlowgladServerAdmin,
  RequestHandlerError: () => import_requestHandler.RequestHandlerError,
  WebhookVerificationError: () => import_webhook.WebhookVerificationError,
  hybridRouteToHandlerMap: () => import_subrouteHandlers.hybridRouteToHandlerMap,
  isHybridActionKey: () => import_subrouteHandlers.isHybridActionKey,
  requestHandler: () => import_requestHandler.requestHandler,
  routeToHandlerMap: () => import_subrouteHandlers.routeToHandlerMap,
  verifyWebhook: () => import_webhook.verifyWebhook
});
module.exports = __toCommonJS(index_exports);
var import_FlowgladServer = require('./FlowgladServer.cjs');
var import_FlowgladServerAdmin = require('./FlowgladServerAdmin.cjs');
var import_requestHandler = require('./requestHandler.cjs');
var import_subrouteHandlers = require('./subrouteHandlers/index.cjs');
var import_webhook = require('./webhook.cjs');
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowgladServer,
  FlowgladServerAdmin,
  RequestHandlerError,
  WebhookVerificationError,
  hybridRouteToHandlerMap,
  isHybridActionKey,
  requestHandler,
  routeToHandlerMap,
  verifyWebhook
});
