// @flowglad/server/dist/esm/FlowgladServerAdmin.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "./chunk-E3LOUS7X.mjs";
import {
  Flowglad as FlowgladNode
} from "@flowglad/node";
import {
  bulkCreateUsageEventsSchema
} from "@flowglad/shared";
class FlowgladServerAdmin {
  constructor(options) {
    this.flowgladNode = new FlowgladNode(options);
  }
  createCustomer(input) {
    return __async(this, null, function* () {
      return this.flowgladNode.customers.create(input);
    });
  }
  getCustomer(externalId) {
    return __async(this, null, function* () {
      return this.flowgladNode.customers.retrieve(externalId);
    });
  }
  getCustomerBilling(externalId) {
    return __async(this, null, function* () {
      return this.flowgladNode.customers.retrieveBilling(externalId);
    });
  }
  updateCustomer(id, input) {
    return __async(this, null, function* () {
      return this.flowgladNode.customers.update(id, input);
    });
  }
  getPricingModel(id) {
    return __async(this, null, function* () {
      return this.flowgladNode.pricingModels.retrieve(id);
    });
  }
  getDefaultPricingModel() {
    return __async(this, null, function* () {
      return this.flowgladNode.pricingModels.retrieveDefault();
    });
  }
  createProduct(input) {
    return __async(this, null, function* () {
      return this.flowgladNode.products.create(input);
    });
  }
  updateProduct(id, input) {
    return __async(this, null, function* () {
      return this.flowgladNode.products.update(id, input);
    });
  }
  clonePricingModel(id, params) {
    return __async(this, null, function* () {
      return this.flowgladNode.pricingModels.clone(id, params);
    });
  }
  getUsageMeter(id) {
    return __async(this, null, function* () {
      return this.flowgladNode.usageMeters.retrieve(id);
    });
  }
  updateUsageMeter(id, params) {
    return __async(this, null, function* () {
      return this.flowgladNode.usageMeters.update(id, params);
    });
  }
  createUsageMeter(params) {
    return __async(this, null, function* () {
      return this.flowgladNode.usageMeters.create(params);
    });
  }
  createUsageEvent(params) {
    return __async(this, null, function* () {
      return this.flowgladNode.usageEvents.create(params);
    });
  }
  /**
   * Create multiple usage events in a single request.
   * NOTE: this method is more efficient than calling `createUsageEvent` multiple times.
   * @param params - The parameters containing an array of usage events.
   * @returns The created usage events.
   */
  bulkCreateUsageEvents(params) {
    return __async(this, null, function* () {
      const parsedParams = bulkCreateUsageEventsSchema.parse(params);
      const usageEvents = parsedParams.usageEvents.map(
        (usageEvent) => {
          var _a, _b;
          return __spreadProps(__spreadValues({}, usageEvent), {
            properties: (_a = usageEvent.properties) != null ? _a : void 0,
            usageDate: (_b = usageEvent.usageDate) != null ? _b : void 0
          });
        }
      );
      return this.flowgladNode.post("/api/v1/usage-events/bulk", {
        body: { usageEvents }
      });
    });
  }
  getUsageEvent(id) {
    return __async(this, null, function* () {
      return this.flowgladNode.usageEvents.retrieve(id);
    });
  }
}
export {
  FlowgladServerAdmin
};
