// @flowglad/server/dist/esm/requestHandler.mjs
import {
  __async
} from "./chunk-E3LOUS7X.mjs";
import {
  FlowgladActionKey
} from "@flowglad/shared";
import { FlowgladServerAdmin } from "./FlowgladServerAdmin.mjs";
import {
  hybridRouteToHandlerMap,
  isHybridActionKey,
  routeToHandlerMap
} from "./subrouteHandlers/index.mjs";
class RequestHandlerError extends Error {
  constructor(message, status = 400) {
    super(message);
    this.status = status;
    this.name = "RequestHandlerError";
  }
}
const requestHandler = (options) => {
  const {
    getCustomerExternalId,
    flowglad,
    apiKey,
    baseURL,
    onError,
    beforeRequest,
    afterRequest
  } = options;
  return (input, request) => __async(null, null, function* () {
    var _a;
    try {
      if (beforeRequest) {
        yield beforeRequest();
      }
      const joinedPath = input.path.join("/");
      if (!Object.values(FlowgladActionKey).includes(joinedPath)) {
        throw new RequestHandlerError(
          `"${joinedPath}" is not a valid Flowglad API path`,
          404
        );
      }
      const data = input.method === "GET" ? input.query : input.body;
      if (isHybridActionKey(joinedPath)) {
        const resolvedApiKey = apiKey || process.env.FLOWGLAD_SECRET_KEY;
        if (!resolvedApiKey) {
          throw new RequestHandlerError(
            "API key required for this route. Provide apiKey option or set FLOWGLAD_SECRET_KEY environment variable.",
            500
          );
        }
        const flowgladServerAdmin = new FlowgladServerAdmin({
          apiKey: resolvedApiKey,
          baseURL: baseURL || process.env.FLOWGLAD_BASE_URL
        });
        let flowgladServer2 = null;
        try {
          const customerExternalId2 = yield getCustomerExternalId(request);
          flowgladServer2 = yield flowglad(customerExternalId2);
        } catch (authError) {
        }
        const handler2 = hybridRouteToHandlerMap[joinedPath];
        const result2 = yield handler2(
          {
            method: input.method,
            data
          },
          {
            flowgladServer: flowgladServer2,
            flowgladServerAdmin
          }
        );
        if (afterRequest) {
          yield afterRequest();
        }
        return {
          status: result2.status,
          data: result2.data,
          error: result2.error
        };
      }
      const handler = routeToHandlerMap[joinedPath];
      if (!handler) {
        throw new RequestHandlerError(
          `"${joinedPath}" is not a valid Flowglad API path`,
          404
        );
      }
      const customerExternalId = yield getCustomerExternalId(request);
      const flowgladServer = yield flowglad(customerExternalId);
      const result = yield handler(
        {
          method: input.method,
          data
        },
        flowgladServer
      );
      if (afterRequest) {
        yield afterRequest();
      }
      return {
        status: result.status,
        data: result.data,
        error: result.error
      };
    } catch (error) {
      if (onError) {
        onError(error);
      }
      if (error instanceof RequestHandlerError) {
        return {
          status: error.status,
          error: { message: error.message }
        };
      }
      if (error.message) {
        return {
          status: (_a = error.status) != null ? _a : 500,
          error: { message: error.message }
        };
      }
      return {
        status: 400,
        error: { message: "Internal server error" }
      };
    }
  });
};
export {
  RequestHandlerError,
  requestHandler
};
