// @flowglad/react/dist/esm/FlowgladContext.js
"use client";
import {
  __async,
  __objRest,
  __spreadProps,
  __spreadValues
} from "./chunk-CRTRMMJ7.js";
import {
  constructCheckFeatureAccess,
  constructCheckUsageBalance,
  constructGetPrice,
  constructGetProduct,
  constructHasPurchased,
  FlowgladActionKey,
  flowgladActionValidators
} from "@flowglad/shared";
import { useQuery, useQueryClient } from "@tanstack/react-query";
import { useMemo } from "react";
import { useFlowgladConfig } from "./FlowgladConfigContext";
import { devError } from "./lib/utils";
import { USAGE_METERS_QUERY_KEY } from "./useUsageMeters";
import { validateUrl } from "./utils";
const getFlowgladRoute = (baseURL, betterAuthBasePath) => {
  var _a;
  if (betterAuthBasePath) {
    const sanitizedPath = betterAuthBasePath.trim().replace(/\/+$/, "");
    return `${sanitizedPath}/flowglad`;
  }
  const sanitizedBaseURL = (_a = baseURL == null ? void 0 : baseURL.trim()) != null ? _a : "";
  if (sanitizedBaseURL !== "") {
    return `${sanitizedBaseURL.replace(/\/+$/, "")}/api/flowglad`;
  }
  return "/api/flowglad";
};
const notPresentContextValues = {
  customer: null,
  subscriptions: null,
  createCheckoutSession: null,
  createAddPaymentMethodCheckoutSession: null,
  createActivateSubscriptionCheckoutSession: null,
  createUsageEvent: null,
  checkFeatureAccess: null,
  checkUsageBalance: null,
  hasPurchased: null,
  pricingModel: null,
  billingPortalUrl: null,
  reload: null,
  catalog: null,
  invoices: [],
  paymentMethods: [],
  purchases: [],
  cancelSubscription: null,
  uncancelSubscription: null,
  adjustSubscription: null,
  currentSubscriptions: [],
  currentSubscription: null
};
const constructCheckoutSessionCreator = (actionKey, baseURL, betterAuthBasePath, requestConfig, mapPayload) => (params) => __async(null, null, function* () {
  var _b;
  validateUrl(params.successUrl, "successUrl");
  validateUrl(params.cancelUrl, "cancelUrl");
  if (baseURL) {
    validateUrl(baseURL, "baseURL", true);
  }
  const headers = requestConfig == null ? void 0 : requestConfig.headers;
  const _a = params, { autoRedirect } = _a, basePayload = __objRest(_a, ["autoRedirect"]);
  const payload = (_b = mapPayload == null ? void 0 : mapPayload(params, basePayload)) != null ? _b : basePayload;
  const flowgladRoute = getFlowgladRoute(
    baseURL,
    betterAuthBasePath
  );
  const response = yield fetch(`${flowgladRoute}/${actionKey}`, {
    method: "POST",
    headers: __spreadValues({
      "Content-Type": "application/json"
    }, headers),
    body: JSON.stringify(payload)
  });
  const json = yield response.json();
  const data = json.data;
  if (json.error) {
    console.error(
      "FlowgladContext: Checkout session creation failed",
      json
    );
    return { error: json.error };
  }
  if (autoRedirect) {
    window.location.href = data.url;
  }
  return { id: data.checkoutSession.id, url: data.url };
});
const constructCancelSubscription = (constructParams) => (params) => __async(null, null, function* () {
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    queryClient
  } = constructParams;
  const headers = requestConfig == null ? void 0 : requestConfig.headers;
  const flowgladRoute = getFlowgladRoute(
    baseURL,
    betterAuthBasePath
  );
  const response = yield fetch(
    `${flowgladRoute}/${FlowgladActionKey.CancelSubscription}`,
    {
      method: "POST",
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, headers),
      body: JSON.stringify(params)
    }
  );
  const json = yield response.json();
  const data = json.data;
  if (json.error) {
    console.error(
      "FlowgladContext: Subscription cancellation failed",
      json
    );
  } else {
    yield queryClient.invalidateQueries({
      queryKey: [FlowgladActionKey.GetCustomerBilling]
    });
  }
  return {
    subscription: data
  };
});
const constructUncancelSubscription = (constructParams) => (params) => __async(null, null, function* () {
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    queryClient
  } = constructParams;
  const headers = requestConfig == null ? void 0 : requestConfig.headers;
  const flowgladRoute = getFlowgladRoute(
    baseURL,
    betterAuthBasePath
  );
  const response = yield fetch(
    `${flowgladRoute}/${FlowgladActionKey.UncancelSubscription}`,
    {
      method: "POST",
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, headers),
      body: JSON.stringify(params)
    }
  );
  const json = yield response.json();
  const data = json.data;
  if (json.error) {
    console.error(
      "FlowgladContext: Subscription uncancellation failed",
      json
    );
  } else {
    yield queryClient.invalidateQueries({
      queryKey: [FlowgladActionKey.GetCustomerBilling]
    });
  }
  return {
    subscription: data
  };
});
const constructAdjustSubscription = (constructParams) => (params) => __async(null, null, function* () {
  var _a;
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    queryClient,
    currentSubscriptions
  } = constructParams;
  const headers = requestConfig == null ? void 0 : requestConfig.headers;
  const flowgladRoute = getFlowgladRoute(
    baseURL,
    betterAuthBasePath
  );
  let subscriptionId = params.subscriptionId;
  if (!subscriptionId) {
    if (!currentSubscriptions || currentSubscriptions.length === 0) {
      throw new Error(
        "No active subscription found for this customer"
      );
    }
    if (currentSubscriptions.length > 1) {
      throw new Error(
        "Customer has multiple active subscriptions. Please specify subscriptionId in params."
      );
    }
    subscriptionId = currentSubscriptions[0].id;
  }
  const response = yield fetch(
    `${flowgladRoute}/${FlowgladActionKey.AdjustSubscription}`,
    {
      method: "POST",
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, headers),
      body: JSON.stringify(__spreadProps(__spreadValues({}, params), {
        subscriptionId
      }))
    }
  );
  if (!response.ok) {
    throw new Error(
      `Subscription adjustment failed: ${response.status} ${response.statusText}`
    );
  }
  const json = yield response.json();
  if (json.error) {
    console.error(
      "FlowgladContext: Subscription adjustment failed",
      json
    );
    throw new Error(
      (_a = json.error.code) != null ? _a : "Subscription adjustment failed"
    );
  }
  if (!json.data) {
    throw new Error(
      "Subscription adjustment failed: no data returned"
    );
  }
  yield queryClient.invalidateQueries({
    queryKey: [FlowgladActionKey.GetCustomerBilling]
  });
  return {
    subscription: json.data
  };
});
const constructCreateUsageEvent = (constructParams) => (params) => __async(null, null, function* () {
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    queryClient
  } = constructParams;
  const headers = requestConfig == null ? void 0 : requestConfig.headers;
  const flowgladRoute = getFlowgladRoute(
    baseURL,
    betterAuthBasePath
  );
  const response = yield fetch(
    `${flowgladRoute}/${FlowgladActionKey.CreateUsageEvent}`,
    {
      method: "POST",
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, headers),
      body: JSON.stringify(params)
    }
  );
  const json = yield response.json();
  if (json.error) {
    console.error(
      "FlowgladContext: Usage event creation failed",
      json
    );
    return { error: json.error };
  }
  yield queryClient.invalidateQueries({
    queryKey: [USAGE_METERS_QUERY_KEY]
  });
  return { usageEvent: { id: json.data.usageEvent.id } };
});
const getFetchImpl = (requestConfig) => {
  var _a;
  const fetchImpl = (_a = requestConfig == null ? void 0 : requestConfig.fetch) != null ? _a : typeof fetch !== "undefined" ? fetch : void 0;
  if (!fetchImpl) {
    throw new Error(
      "fetch is not available. In React Native environments, provide a fetch implementation via requestConfig.fetch"
    );
  }
  return fetchImpl;
};
const hasDataOrError = (value) => {
  if (typeof value !== "object" || value === null) {
    return false;
  }
  return "data" in value || "error" in value;
};
const isCustomerBillingRouteResponse = (value) => {
  return hasDataOrError(value);
};
const isPricingModelRouteResponse = (value) => {
  return hasDataOrError(value);
};
const fetchCustomerBilling = (_0) => __async(null, [_0], function* ({
  baseURL,
  betterAuthBasePath,
  requestConfig
}) {
  const fetchImpl = getFetchImpl(requestConfig);
  const flowgladRoute = getFlowgladRoute(baseURL, betterAuthBasePath);
  const response = yield fetchImpl(
    `${flowgladRoute}/${FlowgladActionKey.GetCustomerBilling}`,
    {
      method: flowgladActionValidators[FlowgladActionKey.GetCustomerBilling].method,
      body: JSON.stringify({}),
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, requestConfig == null ? void 0 : requestConfig.headers)
    }
  );
  try {
    const data = yield response.json();
    if (isCustomerBillingRouteResponse(data)) {
      return data;
    }
    return {
      data: null,
      error: { message: "Unexpected billing response shape" }
    };
  } catch (error) {
    if (process.env.NODE_ENV === "development") {
      console.error("Flowglad: Error fetching billing:", error);
    }
    return {
      data: null,
      error: { message: "Failed to parse billing response JSON" }
    };
  }
});
const fetchPricingModel = (_0) => __async(null, [_0], function* ({
  baseURL,
  betterAuthBasePath,
  requestConfig
}) {
  const fetchImpl = getFetchImpl(requestConfig);
  const flowgladRoute = getFlowgladRoute(baseURL, betterAuthBasePath);
  const response = yield fetchImpl(
    `${flowgladRoute}/${FlowgladActionKey.GetPricingModel}`,
    {
      method: flowgladActionValidators[FlowgladActionKey.GetPricingModel].method,
      body: JSON.stringify({}),
      headers: __spreadValues({
        "Content-Type": "application/json"
      }, requestConfig == null ? void 0 : requestConfig.headers)
    }
  );
  try {
    const data = yield response.json();
    if (isPricingModelRouteResponse(data)) {
      return data;
    }
    return {
      data: null,
      error: {
        code: "UNEXPECTED_PRICING_MODEL_RESPONSE",
        json: {
          message: "Unexpected pricing model response shape"
        }
      }
    };
  } catch (error) {
    if (process.env.NODE_ENV === "development") {
      console.error("Flowglad: Error fetching pricing model:", error);
    }
    return {
      data: null,
      error: {
        code: "PRICING_MODEL_JSON_PARSE_FAILED",
        json: {
          message: "Failed to parse pricing model response JSON"
        }
      }
    };
  }
});
const getDevModeBillingMocks = (billingMocks) => {
  if (!billingMocks) {
    throw new Error(
      "FlowgladProvider: __devMode requires billingMocks"
    );
  }
  return billingMocks;
};
const buildDevModeBillingValue = (billingData) => {
  var _a, _b;
  const getProduct = constructGetProduct(billingData.pricingModel);
  const getPrice = constructGetPrice(billingData.pricingModel);
  const checkFeatureAccess = constructCheckFeatureAccess(
    (_a = billingData.currentSubscriptions) != null ? _a : []
  );
  const checkUsageBalance = constructCheckUsageBalance(
    (_b = billingData.currentSubscriptions) != null ? _b : []
  );
  const hasPurchased = constructHasPurchased(
    billingData.pricingModel,
    billingData.purchases
  );
  return {
    loaded: true,
    errors: null,
    createCheckoutSession: () => Promise.resolve({
      id: "checkout-session-id",
      url: ""
    }),
    createAddPaymentMethodCheckoutSession: () => Promise.resolve({
      id: "checkout-session-id",
      url: ""
    }),
    createActivateSubscriptionCheckoutSession: () => Promise.resolve({
      id: "checkout-session-id",
      url: ""
    }),
    cancelSubscription: (params) => {
      var _a2, _b2, _c, _d;
      const subscription = (_d = (_b2 = (_a2 = billingData.currentSubscriptions) == null ? void 0 : _a2.find(
        (sub) => sub.id === params.id
      )) != null ? _b2 : billingData.currentSubscription) != null ? _d : (_c = billingData.subscriptions) == null ? void 0 : _c.find((sub) => sub.id === params.id);
      if (!subscription) {
        return Promise.reject(
          new Error(
            `Dev mode: no subscription found for id "${params.id}"`
          )
        );
      }
      const now = Date.now();
      return Promise.resolve({
        subscription: {
          subscription: __spreadProps(__spreadValues({}, subscription), {
            current: false,
            status: "canceled",
            canceledAt: now,
            cancelScheduledAt: null,
            updatedAt: now
          })
        }
      });
    },
    uncancelSubscription: (params) => {
      var _a2, _b2, _c, _d;
      const subscription = (_d = (_b2 = (_a2 = billingData.currentSubscriptions) == null ? void 0 : _a2.find(
        (sub) => sub.id === params.id
      )) != null ? _b2 : billingData.currentSubscription) != null ? _d : (_c = billingData.subscriptions) == null ? void 0 : _c.find((sub) => sub.id === params.id);
      if (!subscription) {
        return Promise.reject(
          new Error(
            `Dev mode: no subscription found for id "${params.id}"`
          )
        );
      }
      const now = Date.now();
      return Promise.resolve({
        subscription: {
          subscription: __spreadProps(__spreadValues({}, subscription), {
            current: true,
            status: "active",
            canceledAt: null,
            cancelScheduledAt: null,
            updatedAt: now
          })
        }
      });
    },
    adjustSubscription: (params) => {
      var _a2, _b2, _c, _d, _e;
      let subscriptionId = params.subscriptionId;
      if (!subscriptionId) {
        const currentSubs = (_a2 = billingData.currentSubscriptions) != null ? _a2 : [];
        if (currentSubs.length === 0) {
          return Promise.reject(
            new Error(
              "Dev mode: no active subscription found for this customer"
            )
          );
        }
        if (currentSubs.length > 1) {
          return Promise.reject(
            new Error(
              "Dev mode: customer has multiple active subscriptions. Please specify subscriptionId in params."
            )
          );
        }
        subscriptionId = currentSubs[0].id;
      }
      const subscription = (_e = (_c = (_b2 = billingData.currentSubscriptions) == null ? void 0 : _b2.find(
        (sub) => sub.id === subscriptionId
      )) != null ? _c : billingData.currentSubscription) != null ? _e : (_d = billingData.subscriptions) == null ? void 0 : _d.find(
        (sub) => sub.id === subscriptionId
      );
      if (!subscription) {
        return Promise.reject(
          new Error(
            `Dev mode: no subscription found for id "${subscriptionId}"`
          )
        );
      }
      const now = Date.now();
      return Promise.resolve({
        subscription: {
          subscription: __spreadProps(__spreadValues({}, subscription), {
            updatedAt: now
          }),
          subscriptionItems: [],
          isUpgrade: false,
          resolvedTiming: "immediately"
        }
      });
    },
    createUsageEvent: () => Promise.resolve({
      usageEvent: { id: "dev-usage-event-id" }
    }),
    checkFeatureAccess,
    checkUsageBalance,
    hasPurchased,
    getProduct,
    getPrice,
    reload: () => Promise.resolve(),
    customer: billingData.customer,
    subscriptions: billingData.subscriptions,
    purchases: billingData.purchases,
    invoices: billingData.invoices,
    paymentMethods: billingData.paymentMethods,
    currentSubscription: billingData.currentSubscription,
    currentSubscriptions: billingData.currentSubscriptions,
    billingPortalUrl: billingData.billingPortalUrl,
    pricingModel: billingData.pricingModel,
    // catalog is kept for SDK type compatibility (same data as pricingModel)
    catalog: billingData.catalog
  };
};
const useBilling = () => {
  var _a, _b, _c;
  const queryClient = useQueryClient();
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    __devMode,
    billingMocks
  } = useFlowgladConfig();
  const {
    isPending: isPendingBilling,
    error: errorBilling,
    data: billing
  } = useQuery({
    queryKey: [FlowgladActionKey.GetCustomerBilling],
    enabled: !__devMode,
    queryFn: () => fetchCustomerBilling({
      baseURL,
      betterAuthBasePath,
      requestConfig
    })
  });
  if (__devMode) {
    const billingData = getDevModeBillingMocks(billingMocks);
    return buildDevModeBillingValue(billingData);
  }
  const createCheckoutSession = useMemo(
    () => constructCheckoutSessionCreator(
      FlowgladActionKey.CreateCheckoutSession,
      baseURL,
      betterAuthBasePath,
      requestConfig,
      (_, basePayload) => __spreadProps(__spreadValues({}, basePayload), {
        type: "product"
      })
    ),
    [baseURL, betterAuthBasePath, requestConfig]
  );
  const createAddPaymentMethodCheckoutSession = useMemo(
    () => constructCheckoutSessionCreator(
      FlowgladActionKey.CreateAddPaymentMethodCheckoutSession,
      baseURL,
      betterAuthBasePath,
      requestConfig
    ),
    [baseURL, betterAuthBasePath, requestConfig]
  );
  const createActivateSubscriptionCheckoutSession = useMemo(
    () => constructCheckoutSessionCreator(
      FlowgladActionKey.CreateActivateSubscriptionCheckoutSession,
      baseURL,
      betterAuthBasePath,
      requestConfig
    ),
    [baseURL, betterAuthBasePath, requestConfig]
  );
  const cancelSubscription = useMemo(
    () => constructCancelSubscription({
      baseURL,
      betterAuthBasePath,
      requestConfig,
      queryClient
    }),
    [baseURL, betterAuthBasePath, requestConfig, queryClient]
  );
  const uncancelSubscription = useMemo(
    () => constructUncancelSubscription({
      baseURL,
      betterAuthBasePath,
      requestConfig,
      queryClient
    }),
    [baseURL, betterAuthBasePath, requestConfig, queryClient]
  );
  const createUsageEvent = useMemo(
    () => constructCreateUsageEvent({
      baseURL,
      betterAuthBasePath,
      requestConfig,
      queryClient
    }),
    [baseURL, betterAuthBasePath, requestConfig, queryClient]
  );
  const adjustSubscription = useMemo(
    () => {
      var _a2, _b2;
      return constructAdjustSubscription({
        baseURL,
        betterAuthBasePath,
        requestConfig,
        queryClient,
        currentSubscriptions: (_b2 = (_a2 = billing == null ? void 0 : billing.data) == null ? void 0 : _a2.currentSubscriptions) != null ? _b2 : null
      });
    },
    [
      baseURL,
      betterAuthBasePath,
      requestConfig,
      queryClient,
      (_a = billing == null ? void 0 : billing.data) == null ? void 0 : _a.currentSubscriptions
    ]
  );
  if (billing) {
    const billingError = billing.error;
    const errors2 = [];
    if (billingError) {
      devError(
        `Flowglad route handler error: ${billingError.message}`
      );
      errors2.push(new Error(billingError.message));
    }
    if (billing.data) {
      const billingData = billing.data;
      const reload = () => __async(null, null, function* () {
        yield queryClient.invalidateQueries({
          queryKey: [FlowgladActionKey.GetCustomerBilling]
        });
      });
      const getProduct = constructGetProduct(billingData.pricingModel);
      const getPrice = constructGetPrice(billingData.pricingModel);
      const hasPurchased = constructHasPurchased(
        billingData.pricingModel,
        billingData.purchases
      );
      return {
        loaded: true,
        customer: billingData.customer,
        createCheckoutSession,
        createAddPaymentMethodCheckoutSession,
        cancelSubscription,
        uncancelSubscription,
        adjustSubscription,
        createActivateSubscriptionCheckoutSession,
        createUsageEvent,
        getProduct,
        getPrice,
        hasPurchased,
        checkFeatureAccess: constructCheckFeatureAccess(
          (_b = billingData.currentSubscriptions) != null ? _b : []
        ),
        checkUsageBalance: constructCheckUsageBalance(
          (_c = billingData.currentSubscriptions) != null ? _c : []
        ),
        subscriptions: billingData.subscriptions,
        purchases: billingData.purchases,
        errors: null,
        reload,
        invoices: billingData.invoices,
        paymentMethods: billingData.paymentMethods,
        currentSubscription: billingData.currentSubscription,
        currentSubscriptions: billingData.currentSubscriptions,
        billingPortalUrl: billingData.billingPortalUrl,
        pricingModel: billingData.pricingModel,
        // catalog is kept for SDK type compatibility (same data as pricingModel)
        catalog: billingData.catalog
      };
    }
    if (errors2.length > 0) {
      return __spreadValues({
        loaded: true,
        errors: errors2
      }, notPresentContextValues);
    }
    return __spreadValues({
      loaded: true,
      errors: null
    }, notPresentContextValues);
  }
  if (isPendingBilling) {
    return __spreadValues({
      loaded: false,
      errors: null
    }, notPresentContextValues);
  }
  const errors = [errorBilling].filter(
    (error) => error !== null
  );
  if (errors.length > 0) {
    return __spreadValues({
      loaded: true,
      errors
    }, notPresentContextValues);
  }
  return __spreadValues({
    loaded: true,
    errors: null
  }, notPresentContextValues);
};
const usePricingModel = () => {
  var _a, _b;
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    __devMode,
    billingMocks
  } = useFlowgladConfig();
  const { data: pricingData, error: pricingError } = useQuery({
    queryKey: [FlowgladActionKey.GetPricingModel],
    enabled: !__devMode,
    queryFn: () => fetchPricingModel({
      baseURL,
      betterAuthBasePath,
      requestConfig
    })
  });
  if (__devMode) {
    const billingData = getDevModeBillingMocks(billingMocks);
    return billingData.pricingModel;
  }
  if (pricingError) {
    devError(`Flowglad route handler error: ${pricingError.message}`);
    return null;
  }
  if (pricingData == null ? void 0 : pricingData.error) {
    devError(
      `Flowglad route handler error: ${pricingData.error.code}`
    );
    return null;
  }
  return (_b = (_a = pricingData == null ? void 0 : pricingData.data) == null ? void 0 : _a.pricingModel) != null ? _b : null;
};
const usePricing = () => usePricingModel();
export {
  fetchCustomerBilling,
  fetchPricingModel,
  getFlowgladRoute,
  useBilling,
  usePricing,
  usePricingModel
};
