// @flowglad/react/dist/esm/useUsageMeters.js
"use client";
import {
  __async,
  __spreadValues
} from "./chunk-CRTRMMJ7.js";
import {
  FlowgladActionKey,
  flowgladActionValidators
} from "@flowglad/shared";
import { useQuery } from "@tanstack/react-query";
import { useFlowgladConfig } from "./FlowgladConfigContext";
import { getFlowgladRoute } from "./FlowgladContext";
const USAGE_METERS_QUERY_KEY = "flowglad-usage-meters";
const deriveUsageMeterBalancesFromBillingMocks = (billingMocks) => {
  var _a;
  const currentSubscriptions = (_a = billingMocks.currentSubscriptions) != null ? _a : [];
  const balances = [];
  for (const subscription of currentSubscriptions) {
    const experimental = subscription.experimental;
    if (experimental == null ? void 0 : experimental.usageMeterBalances) {
      balances.push(...experimental.usageMeterBalances);
    }
  }
  return balances;
};
const useUsageMeters = (params) => {
  var _a, _b, _c, _d, _e;
  const {
    baseURL,
    betterAuthBasePath,
    requestConfig,
    __devMode,
    billingMocks
  } = useFlowgladConfig();
  const {
    data: responseData,
    isLoading,
    error
  } = useQuery({
    queryKey: [USAGE_METERS_QUERY_KEY, params == null ? void 0 : params.subscriptionId],
    enabled: !__devMode,
    queryFn: () => __async(null, null, function* () {
      const flowgladRoute = getFlowgladRoute(
        baseURL,
        betterAuthBasePath
      );
      const response = yield fetch(
        `${flowgladRoute}/${FlowgladActionKey.GetUsageMeterBalances}`,
        {
          method: flowgladActionValidators[FlowgladActionKey.GetUsageMeterBalances].method,
          headers: __spreadValues({
            "Content-Type": "application/json"
          }, requestConfig == null ? void 0 : requestConfig.headers),
          body: JSON.stringify(params != null ? params : {})
        }
      );
      const json = yield response.json();
      return json;
    })
  });
  if (__devMode) {
    if (!billingMocks) {
      throw new Error(
        "FlowgladProvider: __devMode requires billingMocks"
      );
    }
    const balances = deriveUsageMeterBalancesFromBillingMocks(billingMocks);
    const filteredBalances = (params == null ? void 0 : params.subscriptionId) ? balances.filter(
      (b) => b.subscriptionId === params.subscriptionId
    ) : balances;
    return {
      usageMeters: filteredBalances,
      isLoading: false,
      error: null
    };
  }
  if (responseData == null ? void 0 : responseData.error) {
    return {
      usageMeters: void 0,
      isLoading: false,
      error: new Error(
        (_d = (_c = (_b = (_a = responseData.error.json) == null ? void 0 : _a.message) == null ? void 0 : _b.toString()) != null ? _c : responseData.error.code) != null ? _d : "Failed to fetch usage meter balances"
      )
    };
  }
  return {
    usageMeters: (_e = responseData == null ? void 0 : responseData.data) == null ? void 0 : _e.usageMeterBalances,
    isLoading,
    error: error != null ? error : null
  };
};
const useUsageMeter = (usageMeterSlug, params) => {
  var _a;
  const { usageMeters, isLoading, error } = useUsageMeters(params);
  const usageMeter = (_a = usageMeters == null ? void 0 : usageMeters.find((b) => b.slug === usageMeterSlug)) != null ? _a : null;
  return { usageMeter, isLoading, error };
};
export {
  USAGE_METERS_QUERY_KEY,
  useUsageMeter,
  useUsageMeters
};
