// @flowglad/server/dist/cjs/better-auth/constants.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  endpointKeyToActionKey: () => endpointKeyToActionKey
});
module.exports = __toCommonJS(constants_exports);
var import_shared = require("@flowglad/shared");
const endpointKeyToActionKey = {
  getCustomerBilling: import_shared.FlowgladActionKey.GetCustomerBilling,
  findOrCreateCustomer: import_shared.FlowgladActionKey.FindOrCreateCustomer,
  createCheckoutSession: import_shared.FlowgladActionKey.CreateCheckoutSession,
  createAddPaymentMethodCheckoutSession: import_shared.FlowgladActionKey.CreateAddPaymentMethodCheckoutSession,
  createActivateSubscriptionCheckoutSession: import_shared.FlowgladActionKey.CreateActivateSubscriptionCheckoutSession,
  cancelSubscription: import_shared.FlowgladActionKey.CancelSubscription,
  uncancelSubscription: import_shared.FlowgladActionKey.UncancelSubscription,
  adjustSubscription: import_shared.FlowgladActionKey.AdjustSubscription,
  createSubscription: import_shared.FlowgladActionKey.CreateSubscription,
  updateCustomer: import_shared.FlowgladActionKey.UpdateCustomer,
  createUsageEvent: import_shared.FlowgladActionKey.CreateUsageEvent,
  getResources: import_shared.FlowgladActionKey.GetResourceUsages,
  getResourceUsage: import_shared.FlowgladActionKey.GetResourceUsage,
  claimResource: import_shared.FlowgladActionKey.ClaimResource,
  releaseResource: import_shared.FlowgladActionKey.ReleaseResource,
  listResourceClaims: import_shared.FlowgladActionKey.ListResourceClaims,
  getUsageMeterBalances: import_shared.FlowgladActionKey.GetUsageMeterBalances
};
const _actionKeyToEndpointKey = {
  [import_shared.FlowgladActionKey.GetCustomerBilling]: "getCustomerBilling",
  [import_shared.FlowgladActionKey.FindOrCreateCustomer]: "findOrCreateCustomer",
  [import_shared.FlowgladActionKey.CreateCheckoutSession]: "createCheckoutSession",
  [import_shared.FlowgladActionKey.CreateAddPaymentMethodCheckoutSession]: "createAddPaymentMethodCheckoutSession",
  [import_shared.FlowgladActionKey.CreateActivateSubscriptionCheckoutSession]: "createActivateSubscriptionCheckoutSession",
  [import_shared.FlowgladActionKey.CancelSubscription]: "cancelSubscription",
  [import_shared.FlowgladActionKey.UncancelSubscription]: "uncancelSubscription",
  [import_shared.FlowgladActionKey.AdjustSubscription]: "adjustSubscription",
  [import_shared.FlowgladActionKey.CreateSubscription]: "createSubscription",
  [import_shared.FlowgladActionKey.UpdateCustomer]: "updateCustomer",
  [import_shared.FlowgladActionKey.CreateUsageEvent]: "createUsageEvent",
  [import_shared.FlowgladActionKey.GetResourceUsages]: "getResources",
  [import_shared.FlowgladActionKey.GetResourceUsage]: "getResourceUsage",
  [import_shared.FlowgladActionKey.ClaimResource]: "claimResource",
  [import_shared.FlowgladActionKey.ReleaseResource]: "releaseResource",
  [import_shared.FlowgladActionKey.ListResourceClaims]: "listResourceClaims",
  [import_shared.FlowgladActionKey.GetUsageMeterBalances]: "getUsageMeterBalances"
};
const _hybridActionKeyToEndpointKey = {
  [import_shared.FlowgladActionKey.GetPricingModel]: "getPricingModel"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  endpointKeyToActionKey
});
