// @flowglad/nextjs/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  FlowgladProvider: () => import_react.FlowgladProvider,
  RESOURCES_QUERY_KEY: () => import_react.RESOURCES_QUERY_KEY,
  RESOURCE_CLAIMS_QUERY_KEY: () => import_react.RESOURCE_CLAIMS_QUERY_KEY,
  USAGE_METERS_QUERY_KEY: () => import_react.USAGE_METERS_QUERY_KEY,
  useBilling: () => import_react.useBilling,
  usePricing: () => import_react.usePricing,
  usePricingModel: () => import_react.usePricingModel,
  useResource: () => import_react.useResource,
  useResources: () => import_react.useResources,
  useUsageMeter: () => import_react.useUsageMeter,
  useUsageMeters: () => import_react.useUsageMeters
});
module.exports = __toCommonJS(index_exports);
var import_react = require("@flowglad/react");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowgladProvider,
  RESOURCES_QUERY_KEY,
  RESOURCE_CLAIMS_QUERY_KEY,
  USAGE_METERS_QUERY_KEY,
  useBilling,
  usePricing,
  usePricingModel,
  useResource,
  useResources,
  useUsageMeter,
  useUsageMeters
});
