// @flowglad/nextjs/dist/cjs/nextRouteHandler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var nextRouteHandler_exports = {};
__export(nextRouteHandler_exports, {
  nextRouteHandler: () => nextRouteHandler
});
module.exports = __toCommonJS(nextRouteHandler_exports);
var import_server = require("@flowglad/server");
var import_server2 = require("next/server");
const nextRouteHandler = (options) => {
  const {
    getCustomerExternalId,
    flowglad,
    onError,
    beforeRequest,
    afterRequest,
    baseURL
  } = options;
  const routeHandler = (_0, _1) => __async(null, [_0, _1], function* (req, {
    params
  }) {
    try {
      const handler = (0, import_server.requestHandler)({
        getCustomerExternalId,
        flowglad,
        onError,
        beforeRequest,
        afterRequest,
        baseURL
      });
      const resolvedParams = "then" in params ? yield params : params;
      const { path } = resolvedParams;
      const result = yield handler(
        {
          path,
          method: req.method,
          query: req.method === "GET" ? Object.fromEntries(req.nextUrl.searchParams) : void 0,
          body: req.method !== "GET" ? yield req.json().catch(() => ({})) : void 0
        },
        req
      );
      return import_server2.NextResponse.json(
        {
          error: result.error,
          data: result.data
        },
        {
          status: result.status
        }
      );
    } catch (error) {
      if (onError) {
        onError(error);
      }
      return import_server2.NextResponse.json(
        {
          error: {
            message: error instanceof Error ? error.message : "Internal server error"
          },
          data: null
        },
        { status: 500 }
      );
    }
  });
  return {
    GET: routeHandler,
    POST: routeHandler
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  nextRouteHandler
});
