// @flowglad/nextjs/dist/cjs/server.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var server_exports = {};
__export(server_exports, {
  FlowgladServer: () => import_server.FlowgladServer,
  FlowgladServerAdmin: () => import_server.FlowgladServerAdmin,
  RequestHandlerError: () => import_server.RequestHandlerError,
  WebhookVerificationError: () => import_server.WebhookVerificationError,
  mcpHandlerWithFlowglad: () => import_withMCPToolAuthorization.mcpHandlerWithFlowglad,
  nextRouteHandler: () => import_nextRouteHandler.nextRouteHandler,
  pagesRouteHandler: () => import_pagesRouteHandler.pagesRouteHandler,
  requestHandler: () => import_server.requestHandler,
  toolWithFeatureAccessCheck: () => import_withMCPToolAuthorization.toolWithFeatureAccessCheck,
  toolWithUsageBalanceCheck: () => import_withMCPToolAuthorization.toolWithUsageBalanceCheck,
  verifyWebhook: () => import_server.verifyWebhook
});
module.exports = __toCommonJS(server_exports);
var import_server = require("@flowglad/server");
var import_nextRouteHandler = require("./nextRouteHandler");
var import_pagesRouteHandler = require("./pagesRouteHandler");
var import_withMCPToolAuthorization = require("./withMCPToolAuthorization");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowgladServer,
  FlowgladServerAdmin,
  RequestHandlerError,
  WebhookVerificationError,
  mcpHandlerWithFlowglad,
  nextRouteHandler,
  pagesRouteHandler,
  requestHandler,
  toolWithFeatureAccessCheck,
  toolWithUsageBalanceCheck,
  verifyWebhook
});
