// @flowglad/nextjs/dist/cjs/withMCPToolAuthorization.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var withMCPToolAuthorization_exports = {};
__export(withMCPToolAuthorization_exports, {
  mcpHandlerWithFlowglad: () => mcpHandlerWithFlowglad,
  toolWithFeatureAccessCheck: () => toolWithFeatureAccessCheck,
  toolWithUsageBalanceCheck: () => toolWithUsageBalanceCheck
});
module.exports = __toCommonJS(withMCPToolAuthorization_exports);
var import_mcp_adapter = require("@vercel/mcp-adapter");
const toolWithFeatureAccessCheck = (toolCallback, params) => {
  const wrappedCallback = (...args) => __async(null, null, function* () {
    const { featureSlug, flowgladServer, upgradePriceSlug } = params;
    const billing = yield flowgladServer.getBilling();
    const feature = billing.checkFeatureAccess(featureSlug);
    if (!feature) {
      const price = billing.getPrice(upgradePriceSlug);
      if (!price) {
        return {
          content: [
            {
              type: "text",
              text: "Upgrade price not found"
            }
          ]
        };
      }
      const checkoutSession = yield params.flowgladServer.createCheckoutSession({
        priceId: price.id,
        quantity: 1,
        successUrl: params.successUrl,
        cancelUrl: params.cancelUrl
      });
      return {
        content: [
          {
            type: "text",
            text: `Feature access denied. To access this feature, please upgrade at: ${checkoutSession.url}`
          }
        ]
      };
    }
    return yield toolCallback(...args);
  });
  return wrappedCallback;
};
const toolWithUsageBalanceCheck = (toolCallback, featureSlug) => {
  return (...args) => __async(null, null, function* () {
    return yield toolCallback(...args);
  });
};
const mcpHandlerWithFlowglad = (constructor, flowgladConstructor) => {
  const handler = (request) => __async(null, null, function* () {
    const flowglad = yield flowgladConstructor(request);
    const mcpHandler = (0, import_mcp_adapter.createMcpHandler)((server) => __async(null, null, function* () {
      yield constructor(server, flowglad);
      return server;
    }));
    return yield mcpHandler(request);
  });
  return handler;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  mcpHandlerWithFlowglad,
  toolWithFeatureAccessCheck,
  toolWithUsageBalanceCheck
});
