// @flowglad/server/dist/esm/express/expressRouter.mjs
import {
  __async
} from "../chunk-E3LOUS7X.mjs";
import { Router } from "express";
import {
  requestHandler
} from "../requestHandler.mjs";
const normalizeQueryParameters = (query) => {
  return Object.fromEntries(
    Object.entries(query).map(([key, value]) => [
      key,
      Array.isArray(value) ? value[0] : typeof value === "object" && value !== null ? null : String(value)
    ]).filter(
      (entry) => entry[1] !== null && entry[1] !== void 0 && entry[1] !== "undefined"
    )
  );
};
const expressRouter = (options) => {
  const router = Router();
  const handler = requestHandler(options);
  if (options.middleware) {
    router.use(...options.middleware);
  }
  router.all("*", (req, res) => __async(null, null, function* () {
    const pathFragments = req.path.split("/").filter((fragment) => fragment !== "");
    const result = yield handler(
      {
        path: pathFragments,
        method: req.method,
        query: req.method === "GET" ? normalizeQueryParameters(req.query) : void 0,
        body: req.method !== "GET" ? req.body : void 0
      },
      req
    );
    res.status(result.status).json({
      error: result.error,
      data: result.data
    });
  }));
  return router;
};
export {
  expressRouter
};
