// @flowglad/server/dist/cjs/supabase/supabaseEdgeHandler.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var supabaseEdgeHandler_exports = {};
__export(supabaseEdgeHandler_exports, {
  supabaseEdgeHandler: () => supabaseEdgeHandler
});
module.exports = __toCommonJS(supabaseEdgeHandler_exports);
var import_requestHandler = require('../requestHandler.cjs');
const extractPath = (pathname) => {
  const trimmedPath = pathname.replace(/^\/+|\/+$/g, "");
  return trimmedPath === "" ? [] : trimmedPath.split("/");
};
const supabaseEdgeHandler = (options) => {
  const handler = (0, import_requestHandler.requestHandler)(options);
  return (req) => __async(null, null, function* () {
    var _a, _b;
    let url;
    try {
      url = new URL(req.url);
    } catch (err) {
      if (options.onError) {
        options.onError(err);
      }
      return new Response(
        JSON.stringify({
          data: null,
          error: { message: "Invalid request URL" }
        }),
        {
          status: 400,
          headers: { "Content-Type": "application/json" }
        }
      );
    }
    let pathSegments = [];
    if (options.basePath) {
      const basePathTrimmed = options.basePath.replace(
        /^\/+|\/+$/g,
        ""
      );
      const basePathWithSlash = `/${basePathTrimmed}`;
      if (url.pathname.startsWith(basePathWithSlash)) {
        const relativePath = url.pathname.slice(
          basePathWithSlash.length
        );
        pathSegments = extractPath(relativePath);
      } else {
        pathSegments = extractPath(url.pathname);
      }
    } else {
      const segments = extractPath(url.pathname);
      if (segments.length >= 3 && segments[0] === "functions" && segments[1] === "v1") {
        pathSegments = segments.slice(3);
      } else {
        pathSegments = segments;
      }
    }
    const queryParams = Array.from(url.searchParams.entries());
    const query = queryParams.length > 0 ? Object.fromEntries(queryParams) : void 0;
    let body = void 0;
    if (req.method !== "GET") {
      try {
        body = yield req.json();
      } catch (e) {
      }
    }
    try {
      const result = yield handler(
        {
          path: pathSegments,
          method: req.method,
          query,
          body
        },
        req
      );
      return new Response(
        JSON.stringify({
          data: (_a = result.data) != null ? _a : null,
          error: (_b = result.error) != null ? _b : null
        }),
        {
          status: result.status,
          headers: {
            "Content-Type": "application/json"
          }
        }
      );
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : "Internal server error";
      return new Response(
        JSON.stringify({
          data: null,
          error: { message: errorMessage }
        }),
        {
          status: 500,
          headers: { "Content-Type": "application/json" }
        }
      );
    }
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  supabaseEdgeHandler
});
