// @flowglad/shared/dist/cjs/actions.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var actions_exports = {};
__export(actions_exports, {
  adjustSubscriptionParamsSchema: () => adjustSubscriptionParamsSchema,
  billingAddressSchema: () => billingAddressSchema,
  bulkCreateUsageEventsSchema: () => bulkCreateUsageEventsSchema,
  cancelSubscriptionSchema: () => cancelSubscriptionSchema,
  claimResourceSchema: () => claimResourceSchema,
  clientCreateUsageEventSchema: () => clientCreateUsageEventSchema,
  createActivateSubscriptionCheckoutSessionSchema: () => createActivateSubscriptionCheckoutSessionSchema,
  createAddPaymentMethodCheckoutSessionSchema: () => createAddPaymentMethodCheckoutSessionSchema,
  createProductCheckoutSessionSchema: () => createProductCheckoutSessionSchema,
  createSubscriptionSchema: () => createSubscriptionSchema,
  createUsageEventSchema: () => createUsageEventSchema,
  flowgladActionValidators: () => flowgladActionValidators,
  getResourceUsageSchema: () => getResourceUsageSchema,
  getResourcesSchema: () => getResourcesSchema,
  getUsageMeterBalancesSchema: () => getUsageMeterBalancesSchema,
  listResourceClaimsSchema: () => listResourceClaimsSchema,
  releaseResourceSchema: () => releaseResourceSchema,
  subscriptionAdjustmentTiming: () => subscriptionAdjustmentTiming,
  subscriptionAdjustmentTimingSchema: () => subscriptionAdjustmentTimingSchema,
  terseSubscriptionItemSchema: () => terseSubscriptionItemSchema,
  uncancelSubscriptionSchema: () => uncancelSubscriptionSchema,
  updateCustomerInputSchema: () => updateCustomerInputSchema,
  updateCustomerSchema: () => updateCustomerSchema
});
module.exports = __toCommonJS(actions_exports);
var import_zod = require("zod");
var import_sdk = require('./types/sdk.cjs');
const createCoreCheckoutSessionSchema = import_zod.z.object({
  successUrl: import_zod.z.url(),
  cancelUrl: import_zod.z.url(),
  outputMetadata: import_zod.z.record(import_zod.z.string(), import_zod.z.any()).optional(),
  outputName: import_zod.z.string().optional()
});
const checkoutSessionWithPriceId = createCoreCheckoutSessionSchema.extend({
  priceId: import_zod.z.string(),
  priceSlug: import_zod.z.never().optional(),
  // Explicitly disallow
  quantity: import_zod.z.number().optional().default(1)
});
const checkoutSessionWithPriceSlug = createCoreCheckoutSessionSchema.extend({
  priceSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  // Explicitly disallow
  quantity: import_zod.z.number().optional().default(1)
});
const createProductCheckoutSessionSchema = import_zod.z.union([
  checkoutSessionWithPriceId,
  checkoutSessionWithPriceSlug
]);
const createAddPaymentMethodCheckoutSessionSchema = createCoreCheckoutSessionSchema.extend({
  targetSubscriptionId: import_zod.z.string().optional()
});
const createActivateSubscriptionCheckoutSessionSchema = createCoreCheckoutSessionSchema.extend({
  targetSubscriptionId: import_zod.z.string()
});
const subscriptionCancellationTiming = {
  AtEndOfCurrentBillingPeriod: "at_end_of_current_billing_period",
  Immediately: "immediately"
};
const cancellationParametersSchema = import_zod.z.discriminatedUnion("timing", [
  import_zod.z.object({
    timing: import_zod.z.literal(
      subscriptionCancellationTiming.AtEndOfCurrentBillingPeriod
    )
  }),
  import_zod.z.object({
    timing: import_zod.z.literal(subscriptionCancellationTiming.Immediately)
  })
]);
const cancelSubscriptionSchema = import_zod.z.object({
  id: import_zod.z.string(),
  cancellation: cancellationParametersSchema
});
const uncancelSubscriptionSchema = import_zod.z.object({
  id: import_zod.z.string()
});
const subscriptionAdjustmentTiming = {
  Immediately: "immediately",
  AtEndOfCurrentBillingPeriod: "at_end_of_current_billing_period",
  Auto: "auto"
};
const subscriptionAdjustmentTimingSchema = import_zod.z.enum([
  subscriptionAdjustmentTiming.Immediately,
  subscriptionAdjustmentTiming.AtEndOfCurrentBillingPeriod,
  subscriptionAdjustmentTiming.Auto
]);
const terseSubscriptionItemWithPriceId = import_zod.z.object({
  priceId: import_zod.z.string(),
  priceSlug: import_zod.z.never().optional(),
  quantity: import_zod.z.number().int().positive().optional().default(1)
});
const terseSubscriptionItemWithPriceSlug = import_zod.z.object({
  priceSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  quantity: import_zod.z.number().int().positive().optional().default(1)
});
const terseSubscriptionItemSchema = import_zod.z.union([
  terseSubscriptionItemWithPriceId,
  terseSubscriptionItemWithPriceSlug
]);
const adjustmentCommonOptions = {
  subscriptionId: import_zod.z.string().optional(),
  timing: subscriptionAdjustmentTimingSchema.optional().default(subscriptionAdjustmentTiming.Auto),
  prorate: import_zod.z.boolean().optional()
};
const adjustSubscriptionByPriceSlugSchema = import_zod.z.object(__spreadValues({
  priceSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  subscriptionItems: import_zod.z.never().optional(),
  quantity: import_zod.z.number().int().positive().optional().default(1)
}, adjustmentCommonOptions));
const adjustSubscriptionByPriceIdSchema = import_zod.z.object(__spreadValues({
  priceId: import_zod.z.string(),
  priceSlug: import_zod.z.never().optional(),
  subscriptionItems: import_zod.z.never().optional(),
  quantity: import_zod.z.number().int().positive().optional().default(1)
}, adjustmentCommonOptions));
const adjustSubscriptionByItemsSchema = import_zod.z.object(__spreadValues({
  subscriptionItems: import_zod.z.array(terseSubscriptionItemSchema).min(1),
  priceId: import_zod.z.never().optional(),
  priceSlug: import_zod.z.never().optional(),
  quantity: import_zod.z.never().optional()
}, adjustmentCommonOptions));
const adjustSubscriptionParamsSchema = import_zod.z.union([
  adjustSubscriptionByPriceSlugSchema,
  adjustSubscriptionByPriceIdSchema,
  adjustSubscriptionByItemsSchema
]);
const baseUsageEventFields = import_zod.z.object({
  amount: import_zod.z.number(),
  subscriptionId: import_zod.z.string(),
  properties: import_zod.z.record(import_zod.z.string(), import_zod.z.unknown()).optional(),
  transactionId: import_zod.z.string(),
  usageDate: import_zod.z.number().optional()
});
const usageEventWithPriceId = baseUsageEventFields.extend({
  priceId: import_zod.z.string(),
  priceSlug: import_zod.z.never().optional(),
  // Explicitly disallow
  usageMeterId: import_zod.z.never().optional(),
  // Explicitly disallow
  usageMeterSlug: import_zod.z.never().optional()
  // Explicitly disallow
});
const usageEventWithPriceSlug = baseUsageEventFields.extend({
  priceSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  // Explicitly disallow
  usageMeterId: import_zod.z.never().optional(),
  // Explicitly disallow
  usageMeterSlug: import_zod.z.never().optional()
  // Explicitly disallow
});
const usageEventWithUsageMeterId = baseUsageEventFields.extend({
  usageMeterId: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  // Explicitly disallow
  priceSlug: import_zod.z.never().optional(),
  // Explicitly disallow
  usageMeterSlug: import_zod.z.never().optional()
  // Explicitly disallow
});
const usageEventWithUsageMeterSlug = baseUsageEventFields.extend({
  usageMeterSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  // Explicitly disallow
  priceSlug: import_zod.z.never().optional(),
  // Explicitly disallow
  usageMeterId: import_zod.z.never().optional()
  // Explicitly disallow
});
const createUsageEventSchema = import_zod.z.union([
  usageEventWithPriceId,
  usageEventWithPriceSlug,
  usageEventWithUsageMeterId,
  usageEventWithUsageMeterSlug
]);
const bulkCreateUsageEventsSchema = import_zod.z.object({
  usageEvents: import_zod.z.array(createUsageEventSchema).min(1)
});
const clientBaseUsageEventFields = import_zod.z.object({
  amount: import_zod.z.number().optional(),
  subscriptionId: import_zod.z.string().optional(),
  properties: import_zod.z.record(import_zod.z.string(), import_zod.z.unknown()).optional(),
  transactionId: import_zod.z.string().optional(),
  usageDate: import_zod.z.number().optional()
});
const clientUsageEventWithPriceId = clientBaseUsageEventFields.extend(
  {
    priceId: import_zod.z.string(),
    priceSlug: import_zod.z.never().optional(),
    usageMeterId: import_zod.z.never().optional(),
    usageMeterSlug: import_zod.z.never().optional()
  }
);
const clientUsageEventWithPriceSlug = clientBaseUsageEventFields.extend({
  priceSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  usageMeterId: import_zod.z.never().optional(),
  usageMeterSlug: import_zod.z.never().optional()
});
const clientUsageEventWithUsageMeterId = clientBaseUsageEventFields.extend({
  usageMeterId: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  priceSlug: import_zod.z.never().optional(),
  usageMeterSlug: import_zod.z.never().optional()
});
const clientUsageEventWithUsageMeterSlug = clientBaseUsageEventFields.extend({
  usageMeterSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional(),
  priceSlug: import_zod.z.never().optional(),
  usageMeterId: import_zod.z.never().optional()
});
const clientCreateUsageEventSchema = import_zod.z.union([
  clientUsageEventWithPriceId,
  clientUsageEventWithPriceSlug,
  clientUsageEventWithUsageMeterId,
  clientUsageEventWithUsageMeterSlug
]);
const createSubscriptionCoreSchema = import_zod.z.object({
  customerId: import_zod.z.string(),
  quantity: import_zod.z.number().optional(),
  startDate: import_zod.z.string().datetime().optional(),
  trialEnd: import_zod.z.number().optional().describe(
    `Epoch time in milliseconds of when the trial ends. If not provided, defaults to startDate + the associated price's trialPeriodDays`
  ),
  metadata: import_zod.z.record(
    import_zod.z.string(),
    import_zod.z.union([import_zod.z.string(), import_zod.z.number(), import_zod.z.boolean()])
  ).optional(),
  name: import_zod.z.string().optional(),
  backupPaymentMethodId: import_zod.z.string().optional(),
  defaultPaymentMethodId: import_zod.z.string().optional(),
  interval: import_zod.z.enum(["day", "week", "month", "year"]).optional(),
  intervalCount: import_zod.z.number().optional(),
  doNotCharge: import_zod.z.boolean().optional().default(false)
});
const createSubscriptionWithPriceId = createSubscriptionCoreSchema.extend({
  priceId: import_zod.z.string(),
  priceSlug: import_zod.z.never().optional()
  // Explicitly disallow
});
const createSubscriptionWithPriceSlug = createSubscriptionCoreSchema.extend({
  priceSlug: import_zod.z.string(),
  priceId: import_zod.z.never().optional()
  // Explicitly disallow
});
const createSubscriptionSchema = import_zod.z.union([
  createSubscriptionWithPriceId,
  createSubscriptionWithPriceSlug
]);
const billingAddressSchema = import_zod.z.object({
  name: import_zod.z.string().optional(),
  firstName: import_zod.z.string().optional(),
  lastName: import_zod.z.string().optional(),
  email: import_zod.z.string().email().optional(),
  address: import_zod.z.object({
    name: import_zod.z.string().optional(),
    line1: import_zod.z.string().nullable(),
    line2: import_zod.z.string().nullable(),
    city: import_zod.z.string().nullable(),
    state: import_zod.z.string().nullable(),
    postal_code: import_zod.z.string().nullable(),
    country: import_zod.z.string()
  }),
  phone: import_zod.z.string().optional()
});
const updateCustomerInputSchema = import_zod.z.object({
  id: import_zod.z.string(),
  name: import_zod.z.string().optional(),
  email: import_zod.z.string().email().optional(),
  phone: import_zod.z.string().optional(),
  billingAddress: billingAddressSchema.optional()
});
const updateCustomerSchema = import_zod.z.object({
  customer: updateCustomerInputSchema,
  externalId: import_zod.z.string()
});
const getResourcesSchema = import_zod.z.object({
  subscriptionId: import_zod.z.string().optional()
});
const claimResourceSchema = import_zod.z.object({
  resourceSlug: import_zod.z.string(),
  subscriptionId: import_zod.z.string().optional(),
  metadata: import_zod.z.record(
    import_zod.z.string(),
    import_zod.z.union([import_zod.z.string().max(500), import_zod.z.number(), import_zod.z.boolean()])
  ).optional(),
  quantity: import_zod.z.number().int().positive().optional().describe(
    "Create N anonymous claims without external identifiers"
  ),
  externalId: import_zod.z.string().max(255).optional().describe("Create a named claim with this external identifier"),
  externalIds: import_zod.z.array(import_zod.z.string().max(255)).nonempty().optional().describe(
    "Create multiple named claims with these external identifiers"
  )
}).refine(
  (data) => {
    const provided = [
      data.quantity,
      data.externalId,
      data.externalIds
    ].filter((v) => v !== void 0);
    return provided.length === 1;
  },
  {
    message: "Exactly one of quantity, externalId, or externalIds must be provided"
  }
);
const releaseResourceSchema = import_zod.z.object({
  resourceSlug: import_zod.z.string(),
  subscriptionId: import_zod.z.string().optional(),
  quantity: import_zod.z.number().int().positive().optional().describe("Release N anonymous claims (FIFO)"),
  externalId: import_zod.z.string().max(255).optional().describe("Release a named claim by external identifier"),
  externalIds: import_zod.z.array(import_zod.z.string().max(255)).nonempty().optional().describe(
    "Release multiple named claims by external identifiers"
  ),
  claimIds: import_zod.z.array(import_zod.z.string()).nonempty().optional().describe("Release specific claims by their IDs")
}).refine(
  (data) => {
    const provided = [
      data.quantity,
      data.externalId,
      data.externalIds,
      data.claimIds
    ].filter((v) => v !== void 0);
    return provided.length === 1;
  },
  {
    message: "Exactly one of quantity, externalId, externalIds, or claimIds must be provided"
  }
);
const listResourceClaimsSchema = import_zod.z.object({
  subscriptionId: import_zod.z.string().optional(),
  resourceSlug: import_zod.z.string().optional()
});
const getResourceUsageSchema = import_zod.z.object({
  subscriptionId: import_zod.z.string().optional()
}).and(
  import_zod.z.union([
    import_zod.z.object({ resourceSlug: import_zod.z.string() }),
    import_zod.z.object({ resourceId: import_zod.z.string() })
  ])
);
const getUsageMeterBalancesSchema = import_zod.z.object({
  subscriptionId: import_zod.z.string().optional()
}).strict();
const flowgladActionValidators = {
  [import_sdk.FlowgladActionKey.GetCustomerBilling]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: import_zod.z.object({
      externalId: import_zod.z.string()
    })
  },
  [import_sdk.FlowgladActionKey.FindOrCreateCustomer]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: import_zod.z.object({
      externalId: import_zod.z.string()
    })
  },
  [import_sdk.FlowgladActionKey.CreateAddPaymentMethodCheckoutSession]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: createAddPaymentMethodCheckoutSessionSchema
  },
  [import_sdk.FlowgladActionKey.CreateActivateSubscriptionCheckoutSession]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: createActivateSubscriptionCheckoutSessionSchema
  },
  [import_sdk.FlowgladActionKey.CreateCheckoutSession]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: createProductCheckoutSessionSchema
  },
  [import_sdk.FlowgladActionKey.CancelSubscription]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: cancelSubscriptionSchema
  },
  [import_sdk.FlowgladActionKey.UncancelSubscription]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: uncancelSubscriptionSchema
  },
  [import_sdk.FlowgladActionKey.AdjustSubscription]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: adjustSubscriptionParamsSchema
  },
  [import_sdk.FlowgladActionKey.CreateSubscription]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: createSubscriptionSchema
  },
  [import_sdk.FlowgladActionKey.UpdateCustomer]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: updateCustomerSchema
  },
  [import_sdk.FlowgladActionKey.CreateUsageEvent]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: clientCreateUsageEventSchema
  },
  [import_sdk.FlowgladActionKey.GetResourceUsages]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: getResourcesSchema
  },
  [import_sdk.FlowgladActionKey.ClaimResource]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: claimResourceSchema
  },
  [import_sdk.FlowgladActionKey.ReleaseResource]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: releaseResourceSchema
  },
  [import_sdk.FlowgladActionKey.ListResourceClaims]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: listResourceClaimsSchema
  },
  [import_sdk.FlowgladActionKey.GetResourceUsage]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: getResourceUsageSchema
  },
  [import_sdk.FlowgladActionKey.GetPricingModel]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: import_zod.z.object({}).strict()
  },
  [import_sdk.FlowgladActionKey.GetUsageMeterBalances]: {
    method: import_sdk.HTTPMethod.POST,
    inputValidator: getUsageMeterBalancesSchema
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  adjustSubscriptionParamsSchema,
  billingAddressSchema,
  bulkCreateUsageEventsSchema,
  cancelSubscriptionSchema,
  claimResourceSchema,
  clientCreateUsageEventSchema,
  createActivateSubscriptionCheckoutSessionSchema,
  createAddPaymentMethodCheckoutSessionSchema,
  createProductCheckoutSessionSchema,
  createSubscriptionSchema,
  createUsageEventSchema,
  flowgladActionValidators,
  getResourceUsageSchema,
  getResourcesSchema,
  getUsageMeterBalancesSchema,
  listResourceClaimsSchema,
  releaseResourceSchema,
  subscriptionAdjustmentTiming,
  subscriptionAdjustmentTimingSchema,
  terseSubscriptionItemSchema,
  uncancelSubscriptionSchema,
  updateCustomerInputSchema,
  updateCustomerSchema
});
