// @flowglad/shared/dist/cjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  FlowgladActionKey: () => import_sdk.FlowgladActionKey,
  HTTPMethod: () => import_sdk.HTTPMethod,
  adjustSubscriptionParamsSchema: () => import_actions.adjustSubscriptionParamsSchema,
  bulkCreateUsageEventsSchema: () => import_actions.bulkCreateUsageEventsSchema,
  cancelSubscriptionSchema: () => import_actions.cancelSubscriptionSchema,
  claimResourceSchema: () => import_actions.claimResourceSchema,
  constructCheckFeatureAccess: () => import_utils.constructCheckFeatureAccess,
  constructCheckUsageBalance: () => import_utils.constructCheckUsageBalance,
  constructGetPrice: () => import_utils.constructGetPrice,
  constructGetProduct: () => import_utils.constructGetProduct,
  constructHasPurchased: () => import_utils.constructHasPurchased,
  createActivateSubscriptionCheckoutSessionSchema: () => import_actions.createActivateSubscriptionCheckoutSessionSchema,
  createAddPaymentMethodCheckoutSessionSchema: () => import_actions.createAddPaymentMethodCheckoutSessionSchema,
  createProductCheckoutSessionSchema: () => import_actions.createProductCheckoutSessionSchema,
  createSubscriptionSchema: () => import_actions.createSubscriptionSchema,
  createUsageEventSchema: () => import_actions.createUsageEventSchema,
  flowgladActionValidators: () => import_actions.flowgladActionValidators,
  getBaseURL: () => import_utils.getBaseURL,
  getResourcesSchema: () => import_actions.getResourcesSchema,
  getUsageMeterBalancesSchema: () => import_actions.getUsageMeterBalancesSchema,
  listResourceClaimsSchema: () => import_actions.listResourceClaimsSchema,
  releaseResourceSchema: () => import_actions.releaseResourceSchema,
  subscriptionAdjustmentTiming: () => import_actions.subscriptionAdjustmentTiming,
  subscriptionAdjustmentTimingSchema: () => import_actions.subscriptionAdjustmentTimingSchema,
  terseSubscriptionItemSchema: () => import_actions.terseSubscriptionItemSchema,
  uncancelSubscriptionSchema: () => import_actions.uncancelSubscriptionSchema,
  updateCustomerSchema: () => import_actions.updateCustomerSchema
});
module.exports = __toCommonJS(index_exports);
var import_actions = require('./actions.cjs');
__reExport(index_exports, require('./types/checkoutSession.cjs'), module.exports);
__reExport(index_exports, require('./types/currency.cjs'), module.exports);
__reExport(index_exports, require('./types/customer.cjs'), module.exports);
__reExport(index_exports, require('./types/invoice.cjs'), module.exports);
__reExport(index_exports, require('./types/payment.cjs'), module.exports);
__reExport(index_exports, require('./types/paymentMethod.cjs'), module.exports);
__reExport(index_exports, require('./types/pricingModel.cjs'), module.exports);
var import_sdk = require('./types/sdk.cjs');
__reExport(index_exports, require('./types/subscription.cjs'), module.exports);
__reExport(index_exports, require('./types/usage.cjs'), module.exports);
var import_utils = require('./utils.cjs');
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FlowgladActionKey,
  HTTPMethod,
  adjustSubscriptionParamsSchema,
  bulkCreateUsageEventsSchema,
  cancelSubscriptionSchema,
  claimResourceSchema,
  constructCheckFeatureAccess,
  constructCheckUsageBalance,
  constructGetPrice,
  constructGetProduct,
  constructHasPurchased,
  createActivateSubscriptionCheckoutSessionSchema,
  createAddPaymentMethodCheckoutSessionSchema,
  createProductCheckoutSessionSchema,
  createSubscriptionSchema,
  createUsageEventSchema,
  flowgladActionValidators,
  getBaseURL,
  getResourcesSchema,
  getUsageMeterBalancesSchema,
  listResourceClaimsSchema,
  releaseResourceSchema,
  subscriptionAdjustmentTiming,
  subscriptionAdjustmentTimingSchema,
  terseSubscriptionItemSchema,
  uncancelSubscriptionSchema,
  updateCustomerSchema,
  ...require('./types/checkoutSession.cjs'),
  ...require('./types/currency.cjs'),
  ...require('./types/customer.cjs'),
  ...require('./types/invoice.cjs'),
  ...require('./types/payment.cjs'),
  ...require('./types/paymentMethod.cjs'),
  ...require('./types/pricingModel.cjs'),
  ...require('./types/subscription.cjs'),
  ...require('./types/usage.cjs')
});
